<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>STAT 117: Data Analysis in Modern Biostatistics – l17-twomarkers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">STAT 117: Data Analysis in Modern Biostatistics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Lectures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures">    
        <li>
    <a class="dropdown-item" href="./L1-intro.html" rel="" target="">
 <span class="dropdown-text">Lecture 1 Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L2-biomarkers.html" rel="" target="">
 <span class="dropdown-text">Lecture 2: Cancer Biomarkers</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L3-data.html" rel="" target="">
 <span class="dropdown-text">Lecture 3: Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L4-metrics.html" rel="" target="">
 <span class="dropdown-text">Lecture 4: Basics of Biomarker Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L5-decision.html" rel="" target="">
 <span class="dropdown-text">Lecture 5 (February 7, 2023) Decision Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L6-bayes.html" rel="" target="">
 <span class="dropdown-text">Lecture 6 (February 9, 2023) Bayesian Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L7-mcmc.html" rel="" target="">
 <span class="dropdown-text">Lecture 7 (February 14, 2023) Monte Carlo Markov Chains</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L8-timetoevent.html" rel="" target="">
 <span class="dropdown-text">Lecture 8 (February 16, 2023) Time to Event Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L9-comparing.html" rel="" target="">
 <span class="dropdown-text">Lecture 9 (February 21, 2023) Comparing Criteria for Biomarker Discovery</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L10-fdr.html" rel="" target="">
 <span class="dropdown-text">Lecture 10 (February 28, 2023) False Discovery Rates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L11-regmean.html" rel="" target="">
 <span class="dropdown-text">Lecture 11: Regression to the Mean</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L12-shrinkage.html" rel="" target="">
 <span class="dropdown-text">Lecture 12: Shrinkage</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L14-shrinkage2.html" rel="" target="">
 <span class="dropdown-text">Lecture 14 (March 21, 2023) Shrinkage part 2</span></a>
  </li>  
        <li class="dropdown-header">L15-metaanlysis2.qmd</li>
        <li>
    <a class="dropdown-item" href="./L17-twomarkers.html" rel="" target="">
 <span class="dropdown-text">Lecture 17 (March 29, 2023)</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#lecture-17-march-29-2023" id="toc-lecture-17-march-29-2023" class="nav-link active" data-scroll-target="#lecture-17-march-29-2023">Lecture 17 (March 29, 2023)</a>
  <ul class="collapse">
  <li><a href="#finding-partners" id="toc-finding-partners" class="nav-link" data-scroll-target="#finding-partners">Finding partners</a>
  <ul class="collapse">
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#one-marker-at-the-time" id="toc-one-marker-at-the-time" class="nav-link" data-scroll-target="#one-marker-at-the-time">One Marker at the Time</a></li>
  </ul></li>
  <li><a href="#two-markers-at-the-time" id="toc-two-markers-at-the-time" class="nav-link" data-scroll-target="#two-markers-at-the-time">Two Markers at the Time</a>
  <ul class="collapse">
  <li><a href="#linear-combination" id="toc-linear-combination" class="nav-link" data-scroll-target="#linear-combination">Linear Combination</a></li>
  <li><a href="#plain-averaging" id="toc-plain-averaging" class="nav-link" data-scroll-target="#plain-averaging">Plain Averaging</a></li>
  <li><a href="#supplementary-figure" id="toc-supplementary-figure" class="nav-link" data-scroll-target="#supplementary-figure">Supplementary Figure</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<div style="page-break-after: always;"></div>
<section id="lecture-17-march-29-2023" class="level1">
<h1>Lecture 17 (March 29, 2023)</h1>
<section id="finding-partners" class="level2">
<h2 class="anchored" data-anchor-id="finding-partners">Finding partners</h2>
<p>Many statistical learning tools acan be used to identify predictors that, as a group, help us classify observations. This lecture covers a baic exercise in model building which, although extremely simplistic compared to even the most common of methods like regularized regression or random forests, can help build intuition for the challenges of building model from high dimensional noisy data. I also use this opportunity to switch to a time-to-event outcome.</p>
<section id="data" class="level3">
<h3 class="anchored" data-anchor-id="data">Data</h3>
<p>In this section we will be utilizing data from the <code>TCGA_eset</code> dataset.</p>
<p>We will be considering continuous biomarker gene expressions in the TCGA study. The phenotype of interest will be a time-to-event variable, <code>days_to_tumor_recurrence</code></p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(survival)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedOvarianData)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(TCGA_eset)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>GGG <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">cbind</span>(<span class="fu">exprs</span>(TCGA_eset)))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>RecurrenceSurvObj <span class="ot">=</span> <span class="fu">Surv</span>( <span class="at">time =</span> <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"days_to_tumor_recurrence"</span>],</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                          <span class="at">event =</span> <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>BinaryRecurrenceAt2y <span class="ot">=</span> <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"days_to_tumor_recurrence"</span>] <span class="sc">&lt;</span> <span class="dv">365</span><span class="sc">*</span><span class="dv">2</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>BinaryRecurrenceAt2y[ <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">!=</span><span class="st">"recurrence"</span> <span class="sc">&amp;</span> <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"days_to_tumor_recurrence"</span>] <span class="sc">&lt;</span> <span class="dv">365</span><span class="sc">*</span><span class="dv">2</span> ] <span class="ot">=</span> <span class="cn">NA</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
<p>We will then filter out the biomarkers with low IQR. A biomarker with a low IQR generally indicates a lack of variation within the range of possible values the biomarker takes. A lack of variation in turn may suggest a lack of discriminative ability. We will filter out genes with an IQR less than or equal to <span class="math inline">\(1\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>IQR.cutoff <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>GGG.IQR <span class="ot">=</span> <span class="fu">apply</span>(GGG,<span class="dv">1</span>,IQR)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>GGG.filtered <span class="ot">=</span> GGG[GGG.IQR<span class="sc">&gt;</span>IQR.cutoff,]</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>GGG.cor <span class="ot">=</span> <span class="fu">cor</span>(<span class="fu">t</span>(GGG.filtered))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="one-marker-at-the-time" class="level3">
<h3 class="anchored" data-anchor-id="one-marker-at-the-time">One Marker at the Time</h3>
<p>Then, we introduce a function that is similar to the <code>CompScores</code> function in preceeding sections. This new function, named <code>CompScoresTE</code>, will help us compute scoring metrics for recurrence based on marginal associations. Specifically, for each biomarker it computes the coefficient resulting from a Cox Proportional Hazards model fit on the biomarker, the negative log p-value of the coefficient, and the IQR of the biomarker.</p>
<div style="page-break-after: always;"></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(survival)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(survcomp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: prodlim</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>CompScoresTE <span class="ot">=</span> <span class="cf">function</span>(XX,YY){</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>NGenes <span class="ot">=</span> <span class="fu">nrow</span>(XX)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>ScoreNames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"coxRR"</span>,<span class="st">"nlpvalue"</span>,<span class="st">"IQR"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>SSS <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="cn">NA</span>,NGenes,<span class="fu">length</span>(ScoreNames)))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SSS) <span class="ot">=</span> ScoreNames</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gg <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>NGenes){</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  cox.out <span class="ot">=</span> <span class="fu">coxph</span>(YY <span class="sc">~</span> XX[gg,])</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"nlpvalue"</span>] <span class="ot">=</span> <span class="sc">-</span><span class="fu">log</span>(<span class="fu">summary</span>(cox.out)<span class="sc">$</span>coefficients[,<span class="st">"Pr(&gt;|z|)"</span>])</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"coxRR"</span>] <span class="ot">=</span> <span class="fu">summary</span>(cox.out)<span class="sc">$</span>coefficients[,<span class="st">"exp(coef)"</span>]</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"IQR"</span>] <span class="ot">=</span> <span class="fu">IQR</span>(XX[gg,])</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(SSS)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>Scores <span class="ot">=</span> <span class="fu">CompScoresTE</span>(<span class="at">XX=</span>GGG.filtered,<span class="at">YY=</span>RecurrenceSurvObj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
<p>From this, we can resample the labels and obtain a null distribution for our data. We then set a cutoff of <span class="math inline">\(1.2\)</span> for the coefficient from the Cox Proportional Hazards model and <span class="math inline">\(7\)</span> for the negative log p-value of the coefficient. We then can tabulate which biomarkers are above these two values, in both the null and observed distributions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>LabelsNull <span class="ot">=</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(GGG.filtered))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>RecurrenceSurvObjNull <span class="ot">=</span>  </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">Surv</span>( <span class="at">time =</span> <span class="fu">pData</span>(TCGA_eset)[LabelsNull,<span class="st">"days_to_tumor_recurrence"</span>], <span class="at">event =</span> <span class="fu">pData</span>(TCGA_eset)[LabelsNull,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>ScoresNull <span class="ot">=</span> <span class="fu">CompScoresTE</span>(<span class="at">XX=</span>GGG.filtered,<span class="at">YY=</span>RecurrenceSurvObjNull)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(Scores[,<span class="st">"coxRR"</span>]<span class="sc">&gt;</span><span class="fl">1.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2180     9 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ScoresNull[,<span class="st">"coxRR"</span>]<span class="sc">&gt;</span><span class="fl">1.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2185     4 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(Scores[,<span class="st">"nlpvalue"</span>]<span class="sc">&gt;</span><span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2183     6 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(ScoresNull[,<span class="st">"nlpvalue"</span>]<span class="sc">&gt;</span><span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 2187     2 </code></pre>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="two-markers-at-the-time" class="level2">
<h2 class="anchored" data-anchor-id="two-markers-at-the-time">Two Markers at the Time</h2>
<section id="linear-combination" class="level3">
<h3 class="anchored" data-anchor-id="linear-combination">Linear Combination</h3>
<p>We will now compute scores for pairs and pick the best partner biomarker. We will also examine all possible additional variables in a multivariable model.</p>
<p>Here one of the two partners is fixed (in this case the winner of the one-gene-at-the-time competition). The score for the pair is the Wald test for the model as a whole (both variables in vs no variable in).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>CompScoresPairs1 <span class="ot">=</span> <span class="cf">function</span>(XX,YY,gg0){</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>NGenes <span class="ot">=</span> <span class="fu">nrow</span>(XX)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>ScoreNames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"coxRR0"</span>,<span class="st">"coxRRa"</span>,<span class="st">"nlpvalueW"</span>,<span class="st">"cor"</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>SSS <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="cn">NA</span>,NGenes,<span class="fu">length</span>(ScoreNames)))</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SSS) <span class="ot">=</span> ScoreNames</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gg <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>NGenes) ){</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  cox.out <span class="ot">=</span> <span class="fu">coxph</span>(YY <span class="sc">~</span> XX[gg,]<span class="sc">+</span>XX[gg0,])</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"coxRR0"</span>] <span class="ot">=</span> <span class="fu">summary</span>(cox.out)<span class="sc">$</span>coefficients[<span class="dv">2</span>,<span class="st">"exp(coef)"</span>]</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"coxRRa"</span>] <span class="ot">=</span> <span class="fu">summary</span>(cox.out)<span class="sc">$</span>coefficients[<span class="dv">1</span>,<span class="st">"exp(coef)"</span>]</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"nlpvalueW"</span>] <span class="ot">=</span> <span class="sc">-</span> <span class="fu">log</span> ( <span class="fu">summary</span>(cox.out)<span class="sc">$</span>wald[<span class="st">"pvalue"</span>] )</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"cor"</span>] <span class="ot">=</span> <span class="fu">cor</span>(XX[gg,],XX[gg0,])</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(SSS)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>TopGene <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Scores))[Scores[,<span class="st">"nlpvalue"</span>]<span class="sc">==</span><span class="fu">max</span>(Scores[,<span class="st">"nlpvalue"</span>])]</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>TopGenes7 <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(Scores))[Scores[,<span class="st">"nlpvalue"</span>]<span class="sc">&gt;</span><span class="dv">7</span>]</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>ScoresPairs1 <span class="ot">=</span> <span class="fu">CompScoresPairs1</span>(<span class="at">XX=</span>GGG.filtered,<span class="at">YY=</span>RecurrenceSurvObj,<span class="at">gg0=</span>TopGene)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>TopPair1 <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ScoresPairs1))[ScoresPairs1[,<span class="st">"nlpvalueW"</span>]<span class="sc">==</span><span class="fu">max</span>(ScoresPairs1[,<span class="st">"nlpvalueW"</span>])]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
<p>We can then plot the single predictor negative log p-value against the negative log p-value associated with multiple variables.</p>
<p>The lone point on the right is the best gene paired with itself, so the horizontal line represents the “bar” we need to meet for the pair to be better than the single.</p>
<p>Colors highlight with positive (blue) or negative (orange) correlation with the top gene, exceeding .1 in absolute value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( Scores[,<span class="st">"nlpvalue"</span>],ScoresPairs1[,<span class="st">"nlpvalueW"</span>], <span class="at">pch=</span><span class="st">"."</span>, <span class="at">cex=</span><span class="dv">5</span>, </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">ylab=</span><span class="st">"Mutivariable negative log p-value"</span>, </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">"Single predictor negative log p-value"</span>  )</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span>Scores[TopGene,<span class="st">"nlpvalue"</span>])</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>PosCor <span class="ot">=</span> ScoresPairs1[ ,<span class="st">"cor"</span>] <span class="sc">&gt;</span> .<span class="dv">1</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>NegCor <span class="ot">=</span> ScoresPairs1[ ,<span class="st">"cor"</span>] <span class="sc">&lt;</span> <span class="sc">-</span> .<span class="dv">1</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( Scores[ PosCor,<span class="st">"nlpvalue"</span>],</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>        ScoresPairs1[ PosCor,<span class="st">"nlpvalueW"</span>], <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( Scores[ NegCor,<span class="st">"nlpvalue"</span>],</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>        ScoresPairs1[ NegCor,<span class="st">"nlpvalueW"</span>], <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"orange"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Multivariable vs.&nbsp;Single Predictor Negative Log P-value Plot</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Comment</p>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="plain-averaging" class="level3">
<h3 class="anchored" data-anchor-id="plain-averaging">Plain Averaging</h3>
<p>We now take a slightly different approach by computing scores by pairs, picking one biomarker, and then looking at all possible “averaging partners”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>CompScoresPairs2 <span class="ot">=</span> <span class="cf">function</span>(XX,YY,gg0){</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>NGenes <span class="ot">=</span> <span class="fu">nrow</span>(XX)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>ScoreNames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"coxRRa"</span>,<span class="st">"nlpvalueW"</span>,<span class="st">"cor"</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>SSS <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="fu">matrix</span>(<span class="cn">NA</span>,NGenes,<span class="fu">length</span>(ScoreNames)))</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(SSS) <span class="ot">=</span> ScoreNames</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gg <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>NGenes) {</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  XXa <span class="ot">=</span> ( XX[gg,]<span class="sc">+</span>XX[gg0,] ) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  cox.out <span class="ot">=</span> <span class="fu">coxph</span>(YY <span class="sc">~</span> XXa)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"coxRRa"</span>] <span class="ot">=</span> <span class="fu">summary</span>(cox.out)<span class="sc">$</span>coefficients[<span class="dv">1</span>,<span class="st">"exp(coef)"</span>]</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"nlpvalueW"</span>] <span class="ot">=</span> <span class="sc">-</span> <span class="fu">log</span> ( <span class="fu">summary</span>(cox.out)<span class="sc">$</span>wald[<span class="st">"pvalue"</span>] )</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  SSS[gg,<span class="st">"cor"</span>] <span class="ot">=</span> <span class="fu">cor</span>(XX[gg,],XX[gg0,])</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="fu">return</span>(SSS)</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>ScoresPairs2 <span class="ot">=</span> <span class="fu">CompScoresPairs2</span>(<span class="at">XX=</span>GGG.filtered,<span class="at">YY=</span>RecurrenceSurvObj,<span class="at">gg0=</span>TopGene)</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>TopPair2 <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ScoresPairs2))[ScoresPairs2[,<span class="st">"nlpvalueW"</span>]<span class="sc">==</span></span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>                                    <span class="fu">max</span>(ScoresPairs2[,<span class="st">"nlpvalueW"</span>])]</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>TopPartners11 <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ScoresPairs2))[ScoresPairs2[,<span class="st">"nlpvalueW"</span>]<span class="sc">&gt;</span><span class="dv">11</span>]</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>TopPartnersNegCor <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ScoresPairs2))[ScoresPairs2[,<span class="st">"nlpvalueW"</span>]<span class="sc">&gt;</span><span class="dv">11</span> <span class="sc">&amp;</span> </span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>                                             ScoresPairs2[ ,<span class="st">"cor"</span>] <span class="sc">&lt;</span> <span class="sc">-</span> .<span class="dv">1</span>]</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a>TopPartnersPosCor <span class="ot">=</span> (<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(ScoresPairs2))[ScoresPairs2[,<span class="st">"nlpvalueW"</span>]<span class="sc">&gt;</span><span class="dv">11</span> <span class="sc">&amp;</span> </span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>                                             ScoresPairs2[ ,<span class="st">"cor"</span>] <span class="sc">&gt;</span> .<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
<p>The plot of interest would then be the single predictor negative log p-value against the averaged predictor negative log p-value.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( Scores[,<span class="st">"nlpvalue"</span>],ScoresPairs2[,<span class="st">"nlpvalueW"</span>], <span class="at">pch=</span><span class="st">"."</span>, <span class="at">cex=</span><span class="dv">3</span>, <span class="at">xlab=</span><span class="st">"Single variable negative log p-value"</span>, <span class="at">ylab=</span><span class="st">"Averaged predictors negative log p-value"</span>  )</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h=</span>Scores[TopGene,<span class="st">"nlpvalue"</span>])</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>PosCor <span class="ot">=</span> ScoresPairs2[ ,<span class="st">"cor"</span>] <span class="sc">&gt;</span> .<span class="dv">1</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( Scores[ PosCor,<span class="st">"nlpvalue"</span>],ScoresPairs2[ PosCor,<span class="st">"nlpvalueW"</span>], </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>NegCor <span class="ot">=</span> ScoresPairs2[ ,<span class="st">"cor"</span>] <span class="sc">&lt;</span> <span class="sc">-</span> .<span class="dv">1</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( Scores[ NegCor,<span class="st">"nlpvalue"</span>],ScoresPairs2[ NegCor,<span class="st">"nlpvalueW"</span>], </span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"orange"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Averaged vs.&nbsp;Single Predictor Negative Log P-value Plot</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Comment:</p>
<ul>
<li><p>Why is the largest p-value encountered in this point and not the previous one?</p></li>
<li><p>Why the lobster claw?</p></li>
<li><p>The second best gene used to be the best partner in the linear combination, but now when we simply average the result is terrible. Why?</p></li>
</ul>
<p>Some hints are in the following two figure, but it helps to puzzle it out for at least a bit before you proceed.</p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>The following table reports the correlation, exponentiated coefficient value, and the associated scores.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(ScoresPairs2[TopGenes7,<span class="fu">c</span>(<span class="st">"cor"</span>,<span class="st">"coxRRa"</span>)],Scores[TopGenes7,<span class="fu">c</span>(<span class="st">"coxRR"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>              cor    coxRRa Scores[TopGenes7, c("coxRR")]
143   0.002878691 1.0729832                     1.1662079
642   1.000000000 0.7770615                     0.7770615
665  -0.079376471 0.9343275                     1.3021766
727   0.066969147 0.6088374                     0.7846375
799   0.179723073 0.7048702                     0.8172432
1529 -0.190374846 1.0129224                     1.2183982</code></pre>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>We can also plot the multivariable negative log p-value against the averaged predictor negative log p-value, further greying out those with marginal positive association with the outcome.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( ScoresPairs1[,<span class="st">"nlpvalueW"</span>],ScoresPairs2[,<span class="st">"nlpvalueW"</span>], <span class="at">pch=</span><span class="st">"."</span>, <span class="at">cex=</span><span class="dv">5</span>, <span class="at">xlab=</span><span class="st">"Mutivariable negative log p-value"</span>, <span class="at">ylab=</span><span class="st">"Averaged predictors negative log p-value"</span> )</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>PosCor <span class="ot">=</span> ScoresPairs1[ ,<span class="st">"cor"</span>] <span class="sc">&gt;</span> .<span class="dv">1</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>NegCor <span class="ot">=</span> ScoresPairs1[ ,<span class="st">"cor"</span>] <span class="sc">&lt;</span> <span class="sc">-</span> .<span class="dv">1</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>PosCoef <span class="ot">=</span> ScoresPairs1[,<span class="st">"coxRRa"</span>] <span class="sc">&gt;</span> <span class="dv">1</span></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( ScoresPairs1[ PosCor,<span class="st">"nlpvalueW"</span>],ScoresPairs2[ PosCor,<span class="st">"nlpvalueW"</span>], </span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"blue"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( ScoresPairs1[ NegCor,<span class="st">"nlpvalueW"</span>],ScoresPairs2[ NegCor,<span class="st">"nlpvalueW"</span>], </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"orange"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>( ScoresPairs1[ PosCoef,<span class="st">"nlpvalueW"</span>],ScoresPairs2[ PosCoef,<span class="st">"nlpvalueW"</span>], </span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">pch=</span><span class="st">"."</span>, <span class="at">col=</span><span class="st">"gray"</span>, <span class="at">cex=</span><span class="dv">5</span> )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Multivariable vs.&nbsp;Averaged Predictor Negative Log P-value Plot</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>An additional plot below compares the top gene and the top pair as identified above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopGene,], GGG.filtered[TopPair1,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Top Pair vs.&nbsp;Top Gene Plot</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="supplementary-figure" class="level3">
<h3 class="anchored" data-anchor-id="supplementary-figure">Supplementary Figure</h3>
<p>A few miscellaneious plots, mostly to give you a sense for the fact that signal is weak and modeling is hard.</p>
<p>We now plot the expression values for the genes with a high combined predictor negative log p-value and that are positively correlated with the identified best gene. The orange/black points indicates recurrence/no recurrence, with recurrence being orange.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">pty=</span><span class="st">"s"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopGene,], GGG.filtered[TopPartnersPosCor[<span class="dv">2</span>],])</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(GGG.filtered[TopGene,], GGG.filtered[TopPartnersPosCor[<span class="dv">2</span>],],<span class="at">col=</span><span class="dv">1</span><span class="sc">+</span><span class="fu">as.vector</span>( <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Scatterplot for High Combined Predictor NLP-value and Positive Correlation with Top Gene</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>NG <span class="ot">=</span> <span class="fu">nrow</span>(GGG.filtered)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>NS <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>))</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">pty=</span><span class="st">"m"</span>)</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GGG.filtered[TopGene,], <span class="fu">rnorm</span>(NS,<span class="dv">0</span>,.<span class="dv">1</span>)<span class="sc">+</span><span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( .<span class="dv">5</span> <span class="sc">*</span> (GGG.filtered[TopGene,]<span class="sc">+</span>GGG.filtered[TopPartnersPosCor[<span class="dv">2</span>],] ), <span class="fu">rnorm</span>(NS,<span class="dv">0</span>,.<span class="dv">1</span>)<span class="sc">+</span><span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>We can also plot the expression values for the genes with a high combined predictor negative log p-value and that are <strong>negatively</strong> correlated with the identified best gene. The red/black points indicates recurrence/no recurrence, with recurrence being red in color.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">pty=</span><span class="st">"s"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopGene,], GGG.filtered[TopPartnersNegCor[<span class="dv">1</span>],])</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(GGG.filtered[TopGene,], GGG.filtered[TopPartnersNegCor[<span class="dv">1</span>],],<span class="at">col=</span><span class="dv">1</span><span class="sc">+</span><span class="fu">as.vector</span>( <span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">pty=</span><span class="st">"m"</span>)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(GGG.filtered[TopGene,], <span class="fu">rnorm</span>(NS,<span class="dv">0</span>,.<span class="dv">1</span>)<span class="sc">+</span><span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( .<span class="dv">5</span> <span class="sc">*</span> (GGG.filtered[TopGene,]<span class="sc">+</span>GGG.filtered[TopPartnersNegCor[<span class="dv">1</span>],] ), <span class="fu">rnorm</span>(NS,<span class="dv">0</span>,.<span class="dv">1</span>)<span class="sc">+</span><span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"recurrence_status"</span>]<span class="sc">==</span><span class="st">"recurrence"</span>) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>The following figures plot the expression values for the multivariable top pairs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopGene,], GGG.filtered[TopPair1,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopPair1,], GGG.filtered[TopPair2,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>We now plot the expression values for the top pairs by averaging.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopGene,], GGG.filtered[TopPair2,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( GGG.filtered[TopPair1,], GGG.filtered[TopPair2,])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>Here are more visualizations through a Kaplan-Meier Curve</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>km.as.one <span class="ot">=</span> <span class="fu">survfit</span> (RecurrenceSurvObj <span class="sc">~</span> <span class="dv">1</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>km.by.topgene <span class="ot">=</span> <span class="fu">survfit</span> (RecurrenceSurvObj <span class="sc">~</span> GGG.filtered[TopGene,] <span class="sc">&gt;</span> <span class="fu">median</span>(GGG.filtered[TopGene,]))</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>km.by.toppair1 <span class="ot">=</span> <span class="fu">survfit</span> (RecurrenceSurvObj <span class="sc">~</span> GGG.filtered[TopPair1,]<span class="sc">&gt;</span></span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(GGG.filtered[TopPair1,]))</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>km.by.toppair2 <span class="ot">=</span> <span class="fu">survfit</span> (RecurrenceSurvObj <span class="sc">~</span> GGG.filtered[TopPair2,]<span class="sc">&gt;</span><span class="fu">median</span>(GGG.filtered[TopPair2,]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(km.as.one)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Kaplan-Meier Curve Regressed on Intercept Only</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(km.by.topgene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Kaplan-Meier Curve Regressed on Top Gene by Median</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(km.by.toppair1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Kaplan-Meier Curve Regressed on Top Pair 1</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(km.by.toppair2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L17-twomarkers_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Kaplan-Meier Curve Regressed on Top Pair 2</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>We can also obtain the Cox Proprotional Hazards Model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>cox.pair1 <span class="ot">=</span> <span class="fu">coxph</span>(RecurrenceSurvObj <span class="sc">~</span> GGG.filtered[TopGene,]<span class="sc">+</span>GGG.filtered[TopPair1,])</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(cox.pair1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
coxph(formula = RecurrenceSurvObj ~ GGG.filtered[TopGene, ] + 
    GGG.filtered[TopPair1, ])

  n= 522, number of events= 266 
   (56 observations deleted due to missingness)

                             coef exp(coef) se(coef)      z Pr(&gt;|z|)    
GGG.filtered[TopGene, ]  -0.24020   0.78647  0.06799 -3.533 0.000411 ***
GGG.filtered[TopPair1, ]  0.14613   1.15735  0.04258  3.432 0.000598 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

                         exp(coef) exp(-coef) lower .95 upper .95
GGG.filtered[TopGene, ]     0.7865      1.272    0.6883    0.8986
GGG.filtered[TopPair1, ]    1.1574      0.864    1.0647    1.2581

Concordance= 0.579  (se = 0.02 )
Likelihood ratio test= 25.72  on 2 df,   p=3e-06
Wald test            = 25.53  on 2 df,   p=3e-06
Score (logrank) test = 25.6  on 2 df,   p=3e-06</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>pair2.mean <span class="ot">=</span> ( GGG.filtered[TopGene,]<span class="sc">+</span>GGG.filtered[TopPair2,] ) <span class="sc">/</span> <span class="dv">2</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>cox.pair2 <span class="ot">=</span> <span class="fu">coxph</span>(RecurrenceSurvObj <span class="sc">~</span> pair2.mean)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(cox.pair2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
coxph(formula = RecurrenceSurvObj ~ pair2.mean)

  n= 522, number of events= 266 
   (56 observations deleted due to missingness)

               coef exp(coef) se(coef)      z Pr(&gt;|z|)    
pair2.mean -0.49620   0.60884  0.09887 -5.019  5.2e-07 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

           exp(coef) exp(-coef) lower .95 upper .95
pair2.mean    0.6088      1.642    0.5016     0.739

Concordance= 0.593  (se = 0.02 )
Likelihood ratio test= 26.28  on 1 df,   p=3e-07
Wald test            = 25.19  on 1 df,   p=5e-07
Score (logrank) test = 25.36  on 1 df,   p=5e-07</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>( <span class="fu">t</span>(GGG.filtered[<span class="fu">c</span>(TopGene,TopPair1,TopPair2),]) )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             FAH         ATF3        FZD4
FAH  1.000000000  0.002878691  0.06696915
ATF3 0.002878691  1.000000000 -0.06247520
FZD4 0.066969147 -0.062475205  1.00000000</code></pre>
</div>
</div>
<!-- Coding that corresponds to L21c.Rmd Ends Here -->


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>