<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.545">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-01-31">

<title>STAT 117: Data Analysis in Modern Biostatistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">STAT 117: Data Analysis in Modern Biostatistics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lectures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures">    
        <li>
    <a class="dropdown-item" href="./L1-intro.html">
 <span class="dropdown-text">Lecture 1 Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L2-biomarkers.html">
 <span class="dropdown-text">Lecture 2: Cancer Biomarkers</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L3-data.html">
 <span class="dropdown-text">Lecture 3: Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L4-metrics.html">
 <span class="dropdown-text">Lecture 4: Basics of Biomarker Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L5-decision.html">
 <span class="dropdown-text">Lecture 5: Decision Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L6-bayes.html">
 <span class="dropdown-text">Lecture 6: Bayesian Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L7-mcmc.html">
 <span class="dropdown-text">Lecture 7: Monte Carlo Markov Chains</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L8-timetoevent.html">
 <span class="dropdown-text">Lecture 8: Time to Event Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L9-comparing.html">
 <span class="dropdown-text">Lecture 9: Comparing Criteria for Biomarker Discovery</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L10-fdr.html">
 <span class="dropdown-text">Lecture 10: False Discovery Rates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L11-regmean.html">
 <span class="dropdown-text">Lecture 11: Regression to the Mean</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L12-shrinkage.html">
 <span class="dropdown-text">Lecture 12: Shrinkage</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L14-shrinkage2.html">
 <span class="dropdown-text">Lecture 14: Shrinkage part 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L15-metaanalysis2.html">
 <span class="dropdown-text">Lecture 15: Metaanalysis part 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L17-twomarkers.html">
 <span class="dropdown-text">Lecture 17</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#lecture-3-data" id="toc-lecture-3-data" class="nav-link active" data-scroll-target="#lecture-3-data">Lecture 3: Data</a>
  <ul class="collapse">
  <li><a href="#high-throughput-gene-expression-measurement" id="toc-high-throughput-gene-expression-measurement" class="nav-link" data-scroll-target="#high-throughput-gene-expression-measurement">High Throughput Gene Expression Measurement</a>
  <ul class="collapse">
  <li><a href="#steps-of-gene-expression-measurement" id="toc-steps-of-gene-expression-measurement" class="nav-link" data-scroll-target="#steps-of-gene-expression-measurement">Steps of Gene Expression Measurement</a></li>
  <li><a href="#affymetrix-microarrays-affy" id="toc-affymetrix-microarrays-affy" class="nav-link" data-scroll-target="#affymetrix-microarrays-affy">Affymetrix Microarrays (“Affy”)</a></li>
  <li><a href="#rna-seq" id="toc-rna-seq" class="nav-link" data-scroll-target="#rna-seq">RNA-seq</a></li>
  </ul></li>
  <li><a href="#batch-effects" id="toc-batch-effects" class="nav-link" data-scroll-target="#batch-effects">Batch Effects</a></li>
  <li><a href="#the-curatedovariandata-package" id="toc-the-curatedovariandata-package" class="nav-link" data-scroll-target="#the-curatedovariandata-package">The CuratedOvarianData package</a>
  <ul class="collapse">
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#preprocessing-of-gene-expression-data-in-curatedovariandata" id="toc-preprocessing-of-gene-expression-data-in-curatedovariandata" class="nav-link" data-scroll-target="#preprocessing-of-gene-expression-data-in-curatedovariandata">Preprocessing of Gene Expression Data in CuratedOvarianData</a></li>
  <li><a href="#clinical-phenotypes-in-the-curatedovariandata-package" id="toc-clinical-phenotypes-in-the-curatedovariandata-package" class="nav-link" data-scroll-target="#clinical-phenotypes-in-the-curatedovariandata-package">Clinical Phenotypes in the CuratedOvarianData Package</a></li>
  <li><a href="#loading-the-package-and-extracting-expression-data-matrices" id="toc-loading-the-package-and-extracting-expression-data-matrices" class="nav-link" data-scroll-target="#loading-the-package-and-extracting-expression-data-matrices">Loading the Package and Extracting Expression Data Matrices</a></li>
  <li><a href="#basic-visualizations" id="toc-basic-visualizations" class="nav-link" data-scroll-target="#basic-visualizations">Basic Visualizations</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">January 31, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="lecture-3-data" class="level1">
<h1>Lecture 3: Data</h1>
<section id="high-throughput-gene-expression-measurement" class="level2">
<h2 class="anchored" data-anchor-id="high-throughput-gene-expression-measurement">High Throughput Gene Expression Measurement</h2>
<section id="steps-of-gene-expression-measurement" class="level3">
<h3 class="anchored" data-anchor-id="steps-of-gene-expression-measurement">Steps of Gene Expression Measurement</h3>
<ul>
<li><p>Specimen Acquisition (get the samples, preserve them between the clinic and the lab, isolate the cells of interest etc)</p></li>
<li><p>Extraction / Preparation Protocols (extract RNA from the samples, often do additional preparation to get it ready for analysis (e.g add tags); often convert RNA into complementary DNA, or cDNA; break sequences into tiny fragments to be reassembled later)</p></li>
<li><p>Quantification (run the experiment)</p></li>
<li><p>Raw Data (typically images)</p></li>
<li><p>Your Data (a heavily preprocessed summary attempting to quantify expression at the aggregate level of a “gene”)</p></li>
</ul>
<div style="page-break-after: always;"></div>
</section>
<section id="affymetrix-microarrays-affy" class="level3">
<h3 class="anchored" data-anchor-id="affymetrix-microarrays-affy">Affymetrix Microarrays (“Affy”)</h3>
<p>We will work with expression data after it has been preprocessed and summarized at the gene label level. We will analyze a lot of data from the Affy platform. Preprocessing is a statistically challenging step. Some of the most important issues for Affymetrix data are discussed by <a href="https://clincancerres.aacrjournals.org/content/clincanres/14/19/5959.full.pdf">@Owzar2008ccr</a></p>
<p>I also highly recommend <a href="https://cbdm.uni-mainz.de/files/2016/02/GE_microarrays.pdf">Katerina Taskova’s Introduction to Microarray Analysis</a> slide deck.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Reading Notes
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>RMA (slides 13-17) was used to preprocess the data we will use. A high-level understanding is useful. It produces summaries of expression at the level of the probe-set.</p></li>
<li><p>It is useful to know the difference between a probe (a single 25-nucleotide sequence), a probe-set (a group of probes predicted to appear on the coding sequence of a gene) and a gene (a region of DNA believed to often be the source of information for the production of a particular set of related proteins).</p></li>
<li><p>Multiple probe-sets can map the same gene (for a digression about this see <span class="citation" data-cites="stalteri2007bmcb">[@stalteri2007bmcb]</span>) To keep things simple and to cross-reference features across technologies we have averaged over these and will use data at the gene level <span class="citation" data-cites="ganzfried2013curatedovariandata">[@ganzfried2013curatedovariandata]</span>.</p></li>
<li><p>Think about quantile normalization (QN) when looking at Slide 19. Does QN help? Does it remove the artifact in bad.cel completely? Does it remove the “fog” in NEG10-1?</p></li>
<li><p>We will cover differential expression later in the course (Slides 23-29). You are welcome to take a peek, but you can skip for now.</p></li>
<li><p>Each of the bullets on Slide 30 embeds its own sources of both noise and systematic biases.</p></li>
</ul>
</div>
</div>
<p>Multiple probe sets for the same gene in Affy are sometimes designed to capture different isoforms (that is messenger RNAs made with different selection of exons). Because Affy needs to position probes near one end of the gene (long story), the extent to which they can capture different isoforms is not great compared to RNA-seq. But there could certainly be exceptions. Affy also used multiple probe sets to increase the number of probes dedicated to a certain genes, when the gene is important or there is uncertainty about mapping. In these cases averaging actually helps.</p>
<p>Rabbit hole: For those interested in a digression on imaging approaches (not covered at all in the class) including a menagerie of artifacts, see <a href="http://dx.doi.org/10.1093/bib/bbm055">@arteaga-salas2007bb</a></p>
<div style="page-break-after: always;"></div>
</section>
<section id="rna-seq" class="level3">
<h3 class="anchored" data-anchor-id="rna-seq">RNA-seq</h3>
<p>RNA-seq is the current alternative to hybridization microarrays. There are several RNA-seq datasets in the CuratedOvarianCanceer and CuratedBreastCancer package you will use in this class, but RNA-seq data exploration will be mostly optional. The main statistical lessons for this course can be learned equally well from Affy data. Of course I would not say this about a Bioinformatics class.</p>
<p>For those interested in understanding RNA-seq a bit more, I recommend the sections “Overview” and “Alignment and Quantification” from <a href="https://doi.org/10.1146/annurev-biodatasci-072018-021255">@VandenBerge2019arbds</a></p>
<p><a href="https://doi.org/10.1186/1471-2105-14-S9-S1">@Xu2013bmcb</a> report a head-to-head comparison of Affy and RNA-seq high throughput measurements. RNA-seq generates count data. It is interesting to look at <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-S9-S1/figures/1">Figure 1</a> and think about how to interpret the “stripes” in panel B.</p>
<p>Microarrays are rarely used for discovery these days. Many of the statistical challenges of biomarker discovery, however, have remained the same. The advantage of using microarrays for this course is there there are far more studies readily available, so certain parts of the course still work a lot better with microarray than RNA-seq.</p>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="batch-effects" class="level2">
<h2 class="anchored" data-anchor-id="batch-effects">Batch Effects</h2>
<p>High-throughput biology data are often produced in batches for efficiency or practical restrictions. Differences across how these batches are handled (in any of the bullets in Slide 30 of <a href="https://cbdm.uni-mainz.de/files/2016/02/GE_microarrays.pdf">Katerina Taskova’s Introduction to Microarray Analysis</a>), can drive significant variation across batches, resulting in so-called batch effects.</p>
<p>CuratedOvarianData has been preprocessed to adjust for batches defined by the study of origin using <a href="https://www.bu.edu/jlab/wp-assets/ComBat/Download.html">Combat</a> <span class="citation" data-cites="Johnson2006">[@Johnson2006]</span>. Other batch effects may be present.</p>
<p><a href="https://doi.org/10.1038/nrg2825">@Leek2010nrg</a> wrote a useful overview of batch effects.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="the-curatedovariandata-package" class="level2">
<h2 class="anchored" data-anchor-id="the-curatedovariandata-package">The CuratedOvarianData package</h2>
<section id="background" class="level3">
<h3 class="anchored" data-anchor-id="background">Background</h3>
<p>Comprehensive review of predictors and datasets for signature validation <a href="https://academic.oup.com/jnci/article-abstract/106/5/dju049/2606979">@Waldron2014</a></p>
<div style="page-break-after: always;"></div>
</section>
<section id="preprocessing-of-gene-expression-data-in-curatedovariandata" class="level3">
<h3 class="anchored" data-anchor-id="preprocessing-of-gene-expression-data-in-curatedovariandata">Preprocessing of Gene Expression Data in CuratedOvarianData</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><embed src="Figures/GanzfriedF1b.pdf" class="img-fluid"></p>
<figcaption>Scheme of preprocessing pipeline for the CuratedOvarianData Package</figcaption>
</figure>
</div>
<p>Detailed description of dataset <a href="https://academic.oup.com/database/article-abstract/doi/10.1093/database/bat013/330978">@ganzfried2013curatedovariandata</a></p>
<div style="page-break-after: always;"></div>
</section>
<section id="clinical-phenotypes-in-the-curatedovariandata-package" class="level3">
<h3 class="anchored" data-anchor-id="clinical-phenotypes-in-the-curatedovariandata-package">Clinical Phenotypes in the CuratedOvarianData Package</h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><embed src="Figures/GanzfriedF2.pdf" class="img-fluid"></p>
<figcaption>This heatmap visualizes for each curated clinical characteristic (rows) the availability in each data set (columns). Red indicates that the corresponding characteristic is available for at least one sample in the data set. See <a href="https://academic.oup.com/database/article-abstract/doi/10.1093/database/bat013/330978">@ganzfried2013curatedovariandata</a> for further details.</figcaption>
</figure>
</div>
<p>It may seem disappointing that there are only two clinical annotations that all of these datasets have. When we originally created the dataset we recontacted all the groups that originally generated the data to try to fill as many of the gaps as possible. Unfortunately some of the clinical data may not have been collected, or may not be easy to share due to privacy concerns or investigators’ reluctance. Since we compiled this data, the field has moved towards larger coordinated efforts across multiple institutions, which have had better outcomes in terms of completeness and homogeneity of phenotypes. An example is <a href="https://dcc.icgc.org/">PCAWG</a> but heterogeneity and missingness of labels remain important issues in health data.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="loading-the-package-and-extracting-expression-data-matrices" class="level3">
<h3 class="anchored" data-anchor-id="loading-the-package-and-extracting-expression-data-matrices">Loading the Package and Extracting Expression Data Matrices</h3>
<p>We next load the CuratedOvarianData package and two datasets contained within. For now we will only consider the datasets <code>TCGA_eset</code> and <code>GSE32063_eset</code>. To check important characteristics of each of the studies, just enter the study ID (for example GSE32063) in the <a href="https://www.ncbi.nlm.nih.gov/geo/">Gene Expression Omnibus query page</a></p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedOvarianData)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="co"># data(package="curatedOvarianData")</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Above line would lists all datasets available within the package.</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(TCGA_eset)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GSE32063_eset)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div style="page-break-after: always;"></div>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">exprs</span>(TCGA_eset)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      TCGA.20.0987 TCGA.23.1031 TCGA.24.0979 TCGA.23.1117 TCGA.23.1021
A1CF      2.923522     3.052169     2.846371     3.002209     3.062993
A2M      10.353008    11.635772     7.954542     9.971500     8.971334
A4GNT     3.321405     3.666463     3.258038     3.596212     3.388706
AAAS      4.608010     5.142133     5.025422     5.139928     5.256831
AACS      7.279213     7.048869     7.750161     6.206031     7.835422
AADAC     4.605331     5.775611     3.846412     4.468379     4.415817
      TCGA.04.1337
A1CF      2.974734
A2M       9.042876
A4GNT     3.269979
AAAS      4.667723
AACS      6.763047
AADAC     4.159804</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">exprs</span>(GSE32063_eset)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          GSM795125 GSM795126 GSM795127  GSM795128 GSM795129  GSM795130
A1BG     -6.3146850 -4.566943 -6.998341 -5.6383615 -6.780428 -5.1891665
A1BG-AS1 -3.6973042 -2.848625 -2.519599 -3.4870706 -3.043527 -2.6745200
A1CF     -5.5054617 -3.861939 -4.552987 -4.5571017 -4.753347 -0.3129711
A2M       3.2657738  3.373391  5.217385  4.3027153  3.599464  3.2226200
A2ML1    -0.8923645 -0.733676 -1.512566 -0.7562313 -1.414893 -0.8385077
A3GALT2  -5.0457100 -4.183009 -4.553020 -4.2992754 -4.304550 -4.4622746</code></pre>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="basic-visualizations" class="level3">
<h3 class="anchored" data-anchor-id="basic-visualizations">Basic Visualizations</h3>
<p>Clustered heatmap of the top 100 most variable genes in the first 50 samples in TCGA. The expression scale is the log of the pre-processed intensities.</p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>sds <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(TCGA_eset)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>],<span class="dv">1</span>,sd)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>((<span class="fu">exprs</span>(TCGA_eset)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>])[<span class="fu">order</span>(sds,<span class="at">decreasing=</span><span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>],],</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>          <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">10</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<p>Trees (dendrograms) on the top and sides represent the result of hierarchical clustering. The goal is to reorder rows and columns so that similar rows (column) are close to each other. In this case I am using them to facilitate visualization, but this is also a technique to identify patterns in the data or divide them into subgroups. This is a general overview <a href="https://biostats.bepress.com/jhubiostat/paper70/">@Garrett2005</a></p>
<p>The heatmap.2 function calls hclust to produce the dendrograms. The procedure is agglomerative —it starts by joining single rows, and proceeds iteratively from the bottom up. The lenth of the branches grows with the distance of the two entities (rows of clusters of rows) being joined. The default agglomeration method is the “complete linkage” method in which clusters are linked based on the similarity of the furthest members. Same for columns.</p>
<p>Blue lines. Inside the legend: histogram frequencies; inside the heat map, the distance of the line from the center of each color-cell is proportional to the size of the measurement (overkill? depends, people perceive location differently from color so it could help). </p><div style="page-break-after: always;"></div><p></p>
<p>Distribution of location (median) and scale (Interquartile Range, or IQR), by gene and by sample.</p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>iqr.g <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(TCGA_eset),<span class="dv">1</span>,IQR) <span class="co">#</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iqr.g, <span class="at">main =</span><span class="st">"IQR by gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>iqr.s <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(TCGA_eset),<span class="dv">2</span>,IQR) </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iqr.s, <span class="at">main =</span><span class="st">"IQR by sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>med.g <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(TCGA_eset),<span class="dv">1</span>,median)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(med.g, <span class="at">main =</span><span class="st">"median by gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>med.s <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(TCGA_eset),<span class="dv">2</span>,median)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(med.s, <span class="at">main =</span><span class="st">"median by sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Quantile Normlization
</div>
</div>
<div class="callout-body-container callout-body">
<p>If quantile normalization was applied by sample, why aren’t all medians and all IQR’s the same in the histograms by sample?</p>
<p>At least two reasons: 1) QN was applied at the proble level. Within each study data sere subsequently aggregated first by probeset and then by gene. 2) To (hopefully) increase comparabiity across studies, data where further process with a batch adjustment model that regresses out gene specific effects.</p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>Clustered heatmap of the top 100 most variable genes in the Yoshirara study. The technology is Agilent. <a href="https://doi.org/10.1158/1078-0432.CCR-11-2725">@yoshihara2012</a> describes the preprocessing approach. This set of 40 patients is “Dataset B” in <span class="citation" data-cites="yoshihara2012">[@yoshihara2012]</span>.</p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sds <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(GSE32063_eset),<span class="dv">1</span>,sd)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap.2</span>(<span class="fu">exprs</span>(GSE32063_eset)[<span class="fu">order</span>(sds,<span class="at">decreasing=</span><span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>],])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>Distribution of location (median) and scale (Interquartile Range, or IQR), by gene and by sample.</p>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>iqr.g <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(GSE32063_eset),<span class="dv">1</span>,IQR)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iqr.g, <span class="at">main =</span><span class="st">"IQR by gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>iqr.s <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(GSE32063_eset),<span class="dv">2</span>,IQR) </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(iqr.s, <span class="at">main =</span><span class="st">"IQR by sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-6-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>med.g <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(GSE32063_eset),<span class="dv">1</span>,median)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(med.g, <span class="at">main =</span><span class="st">"median by gene"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-6-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>med.s <span class="ot">=</span> <span class="fu">apply</span>(<span class="fu">exprs</span>(GSE32063_eset),<span class="dv">2</span>,median)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(med.s, <span class="at">main =</span><span class="st">"median by sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L3-data_files/figure-html/unnamed-chunk-6-4.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<div id="refs" role="list">

</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>