<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-02-02">

<title>STAT 117: Data Analysis in Modern Biostatistics – l4-metrics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">STAT 117: Data Analysis in Modern Biostatistics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html" rel="" target="">
 <span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html" rel="" target="">
 <span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Lectures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures">    
        <li>
    <a class="dropdown-item" href="./L1-intro.html" rel="" target="">
 <span class="dropdown-text">Lecture 1 (January 24, 2023) Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L2-biomarkers.html" rel="" target="">
 <span class="dropdown-text">Lecture 2: Cancer Biomarkers</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L3-data.html" rel="" target="">
 <span class="dropdown-text">Lecture 3: Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L4-metrics.html" rel="" target="">
 <span class="dropdown-text">Lecture 4: Basics of Biomarker Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L5-decision.html" rel="" target="">
 <span class="dropdown-text">Lecture 5 (February 7, 2023) Decision Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L6-bayes.html" rel="" target="">
 <span class="dropdown-text">Lecture 6 (February 9, 2023) Bayesian Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L7-mcmc.html" rel="" target="">
 <span class="dropdown-text">Lecture 7 (February 14, 2023) Monte Carlo Markov Chains</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L8-timetoevent.html" rel="" target="">
 <span class="dropdown-text">Lecture 8 (February 16, 2023) Time to Event Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L9-comparing.html" rel="" target="">
 <span class="dropdown-text">Lecture 9 (February 21, 2023) Comparing Criteria for Biomarker Discovery</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L10-fdr.html" rel="" target="">
 <span class="dropdown-text">Lecture 10 (February 28, 2023) False Discovery Rates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L11-regmean.html" rel="" target="">
 <span class="dropdown-text">Lecture 11: Regression to the Mean</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L12-shrinkage.html" rel="" target="">
 <span class="dropdown-text">Lecture 12: Shrinkage</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L14-shrinkage2.html" rel="" target="">
 <span class="dropdown-text">Lecture 14 (March 21, 2023) Shrinkage part 2</span></a>
  </li>  
        <li class="dropdown-header">L15-metaanlysis2.qmd</li>
        <li>
    <a class="dropdown-item" href="./L17-twomarkers.html" rel="" target="">
 <span class="dropdown-text">Lecture 17 (March 29, 2023)</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#lecture-4-basics-of-biomarker-evaluation" id="toc-lecture-4-basics-of-biomarker-evaluation" class="nav-link active" data-scroll-target="#lecture-4-basics-of-biomarker-evaluation">Lecture 4: Basics of Biomarker Evaluation</a>
  <ul class="collapse">
  <li><a href="#znf487-and-debulking-in-yoshihara-study-b" id="toc-znf487-and-debulking-in-yoshihara-study-b" class="nav-link" data-scroll-target="#znf487-and-debulking-in-yoshihara-study-b">ZNF487 and debulking in Yoshihara Study B</a></li>
  <li><a href="#visualizations-of-biomarker-distribution-conditional-on-outcome" id="toc-visualizations-of-biomarker-distribution-conditional-on-outcome" class="nav-link" data-scroll-target="#visualizations-of-biomarker-distribution-conditional-on-outcome">Visualizations of Biomarker Distribution Conditional on Outcome</a>
  <ul class="collapse">
  <li><a href="#dot-plots" id="toc-dot-plots" class="nav-link" data-scroll-target="#dot-plots">Dot Plots</a></li>
  <li><a href="#conditional-density-plots" id="toc-conditional-density-plots" class="nav-link" data-scroll-target="#conditional-density-plots">Conditional Density Plots</a></li>
  <li><a href="#positive-predictive-value-curves" id="toc-positive-predictive-value-curves" class="nav-link" data-scroll-target="#positive-predictive-value-curves">Positive Predictive Value Curves</a></li>
  <li><a href="#log-likelihood-ratio-curve" id="toc-log-likelihood-ratio-curve" class="nav-link" data-scroll-target="#log-likelihood-ratio-curve">(log) Likelihood Ratio Curve</a></li>
  </ul></li>
  <li><a href="#threshold-based-classification" id="toc-threshold-based-classification" class="nav-link" data-scroll-target="#threshold-based-classification">Threshold-based Classification</a>
  <ul class="collapse">
  <li><a href="#x2-tables" id="toc-x2-tables" class="nav-link" data-scroll-target="#x2-tables">2x2 tables</a></li>
  <li><a href="#visualizing-properties-of-alternative-thresholds" id="toc-visualizing-properties-of-alternative-thresholds" class="nav-link" data-scroll-target="#visualizing-properties-of-alternative-thresholds">Visualizing Properties of Alternative Thresholds</a></li>
  <li><a href="#the-roc-curve" id="toc-the-roc-curve" class="nav-link" data-scroll-target="#the-roc-curve">The ROC curve</a></li>
  <li><a href="#the-precision-recall-curve" id="toc-the-precision-recall-curve" class="nav-link" data-scroll-target="#the-precision-recall-curve">The Precision / Recall curve</a></li>
  <li><a href="#the-det-curve" id="toc-the-det-curve" class="nav-link" data-scroll-target="#the-det-curve">The DET curve</a></li>
  </ul></li>
  <li><a href="#comparing-znf487-and-igha1" id="toc-comparing-znf487-and-igha1" class="nav-link" data-scroll-target="#comparing-znf487-and-igha1">Comparing ZNF487 and IGHA1</a>
  <ul class="collapse">
  <li><a href="#ppv-comparison" id="toc-ppv-comparison" class="nav-link" data-scroll-target="#ppv-comparison">PPV comparison</a></li>
  <li><a href="#roc-comparison" id="toc-roc-comparison" class="nav-link" data-scroll-target="#roc-comparison">ROC comparison</a></li>
  <li><a href="#pr-comparison" id="toc-pr-comparison" class="nav-link" data-scroll-target="#pr-comparison">PR comparison</a></li>
  <li><a href="#det-comparison" id="toc-det-comparison" class="nav-link" data-scroll-target="#det-comparison">DET comparison</a></li>
  <li><a href="#the-skill-plot" id="toc-the-skill-plot" class="nav-link" data-scroll-target="#the-skill-plot">The Skill Plot</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<div style="page-break-after: always;"></div>
<section id="lecture-4-basics-of-biomarker-evaluation" class="level1">
<h1>Lecture 4: Basics of Biomarker Evaluation</h1>
<p><a href="https://doi.org/10.1016/j.molonc.2014.10.004">Biomarker validation and testing</a></p>
<p>The lecture <a href="https://youtu.be/bGH870J3MMM">Biomarkers in Cancer Research</a> by Nabihah Tayob is highly recommended, particularly if you are encountering concepts like positive predictive value or ROC curves for the first time. This lecture also give a bit more context of biomarkers in cancer compared to what we have been able to cover in Lecture 2.</p>
<section id="znf487-and-debulking-in-yoshihara-study-b" class="level2">
<h2 class="anchored" data-anchor-id="znf487-and-debulking-in-yoshihara-study-b">ZNF487 and debulking in Yoshihara Study B</h2>
<p>We first focus on the gene <code>ZNF487</code> contained within the <code>GSE32063_eset</code> dataset. <code>ZNF487</code> stands for “Zinc Finger Protein 487”. The <code>exprs</code> function allows us to pull out expression-level data, while <code>pData</code> retrieves the phenotype data, which are the clinical characteristics and outcomes of patients. In the next chunk of code, we will consider the <code>debulking</code> variable as our response variable, and the expression data for the <code>ZNF487</code> gene as our predictor variable, or biomarker.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Debulking
</div>
</div>
<div class="callout-body-container callout-body">
<p>The term debulking captures whether the patient’s surgery was successful in removing the tumor mass (optimal debulking) in contrast to the case where additional tumor cells where thought to be present based on the surgeon’s inspection, but could not be removed (suboptimal debulking).</p>
<p>A discussion of the clinical relevance of this question is in the introduction of <a href="https://doi.org/10.1093/jnci/dju048">@Riester2014</a>. A useful debulking marker could spare an unnecessary surgery to patients who will not benefit. However, a high degree of reliability would be needed for the decision to forego surgery, as the benefit of a hypotetical successful surgery could be substantial.</p>
</div>
</div>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedOvarianData)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GSE32063_eset)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">pData</span>(GSE32063_eset)[,<span class="st">"debulking"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   optimal suboptimal 
        19         21 </code></pre>
</div>
</div>
<p>We will further define the debulking variable to be <span class="math inline">\(1\)</span> if optimal, and <span class="math inline">\(0\)</span> if suboptimal. For <code>ZNF487</code> this results in higher levels of gene expression being generally associated with optimal debulking.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="visualizations-of-biomarker-distribution-conditional-on-outcome" class="level2">
<h2 class="anchored" data-anchor-id="visualizations-of-biomarker-distribution-conditional-on-outcome">Visualizations of Biomarker Distribution Conditional on Outcome</h2>
<section id="dot-plots" class="level3">
<h3 class="anchored" data-anchor-id="dot-plots">Dot Plots</h3>
<p>A most basic way to begin visualization is a pair of <strong>dot-plots</strong> depicting <em>class-conditional</em> distributions of expression. It allows one to visually gauge the overall separation between expression levels corresponding to optimal versus suboptimal debulking status. The plot appears to indicate that larger and more positive expression values of the biomarker are associated with a debulking status of <span class="math inline">\(1\)</span>, or optimal.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>GeneName <span class="ot">=</span> <span class="st">"ZNF487"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>XX <span class="ot">=</span> <span class="fu">exprs</span>(GSE32063_eset)[GeneName,]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>YY <span class="ot">=</span> <span class="fu">pData</span>(GSE32063_eset)[,<span class="st">"debulking"</span>] <span class="sc">==</span> <span class="st">"optimal"</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(XX,YY,<span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression of"</span>,GeneName),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Debulking Status"</span>,<span class="at">cex=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Dotplot for ZNF487</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Before you go any further pause to think about what you learned from the dot plot about the potential for <code>ZNF487</code> to serve as a useful biomarker.</p>
<p>Think about the question in the abstract and in the context of the surgery decision described above.</p>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="conditional-density-plots" class="level3">
<h3 class="anchored" data-anchor-id="conditional-density-plots">Conditional Density Plots</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ffx <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>x</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>ff0 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>ff1 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ffx,ff0,<span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression"</span>,GeneName),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Density"</span>,<span class="at">main =</span> <span class="st">""</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), <span class="at">col =</span> <span class="st">"blue"</span>,<span class="at">type=</span><span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(ffx,ff1, <span class="at">col =</span> <span class="st">"green"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">rug</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">ticksize =</span> .<span class="dv">05</span>,<span class="at">lwd=</span><span class="dv">2</span>, <span class="at">col =</span> <span class="st">"blue"</span>)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">rug</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>],<span class="at">ticksize =</span> .<span class="dv">05</span>,<span class="at">lwd=</span><span class="dv">2</span>, <span class="at">col =</span> <span class="st">"green"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Density Plot for ZNF487 expression</figcaption>
</figure>
</div>
</div>
</div>
<p>We can also visualize the data via the empirical density curves of expression for different group by debulking status. The green curve is associated with optimal debulking, and the blue with suboptimal debulking. The density plots above are essentially capturing the same variation as the dot-plot. The marks at the bottom correspond to the dotplots points. The density plot provides a different view of the overall variation, particularly in the area of overlap. Generally, multiple visualizations of the same data may assist in better understanding properties of the data.</p>
<p>With small sample sizes this visualization will be very sensitive to the tuning parameters of the density estimate (try the “adjust” input which controls the smoothness of the curve). That said if you have a decent guess about the smoothness of the density, you can use the estimated curves to get various types of useful probabilistic predictions and estimates. Some are coming next.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="positive-predictive-value-curves" class="level3">
<h3 class="anchored" data-anchor-id="positive-predictive-value-curves">Positive Predictive Value Curves</h3>
<p>The previous sets of plots start to indicate that the ZNF487 biomarker may be informative for the purpose of predicting debulking status. Clinically, a biomarker may be used to classify a new patient after having measured her specific level. We are then interested in understanding the or <code>ppv</code> associated with the measured expression, that is the conditional probability of the label given the observed biomarker level. Bayes’ rule allows us to calculate this conditional (or personalized) probability if we know the frequency of optimally debulkable tumors in a population of reference (prevalence) of which the patient in question is a random representative.</p>
<p>The expression measurements we used here are obtained from a fresh frozen tumor sample taken at the time of surgery. If we discovered a usefl biomarker, we would also try to measure it differently, for example based on a less invasive procedure such as a biopsy (wich uses a needle to remove a small section of the tumor). To think about the utility of a marker, imagine it is available before the surgery is performed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>prev <span class="ot">=</span> .<span class="dv">1</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>ppv <span class="ot">=</span> prev <span class="sc">*</span> ff1 <span class="sc">/</span> ( prev <span class="sc">*</span> ff1 <span class="sc">+</span> (<span class="dv">1</span><span class="sc">-</span>prev) <span class="sc">*</span> ff0)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ffx,ppv,<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression"</span>,GeneName),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Probability of optimal debulking"</span>,<span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>prev <span class="ot">=</span> .<span class="dv">2</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>ppv2 <span class="ot">=</span> prev <span class="sc">*</span> ff1 <span class="sc">/</span> ( prev <span class="sc">*</span> ff1 <span class="sc">+</span> (<span class="dv">1</span><span class="sc">-</span>prev) <span class="sc">*</span> ff0) <span class="co"># BAYES RULE</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(ffx,ppv2,<span class="at">lty=</span><span class="st">"dotted"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">PPV Plot for ZNF487 expression</figcaption>
</figure>
</div>
</div>
</div>
<p>The code above computes <code>ppv</code> and graphs it with prevalence set at <code>prev = .1</code> and <span class="math inline">\(.2\)</span>. The dotted line corresponds to a prevalence of <span class="math inline">\(.2\)</span>.</p>
<div style="page-break-after: always;"></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discussion Points:
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p>Which shape do you wish a <code>ppv</code> curve to have for a predictive biomarker to be useful?</p></li>
<li><p>The Yoshihara study B includes about the same number of optimal and suboptimal cases. Why would we be interested in evaluating PPV at prevalences of 10 and 20 percent?</p></li>
<li><p>(related to 2.) Yoshihara et al describe the combined collection of studies A and B by saying “patients who were diagnosed with advanced stage high-grade serous ovarian cancer between July 1997 and June 2010 were included in this study”. Below is breakdown of optimal and suboptimal for study A. Comment.</p></li>
</ol>
</div>
</div>
<div class="cell" data-warnings="false">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedOvarianData)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GSE32062.GPL6480_eset)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">pData</span>(GSE32062.GPL6480_eset)[,<span class="st">"debulking"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   optimal suboptimal 
       103        157 </code></pre>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="log-likelihood-ratio-curve" class="level3">
<h3 class="anchored" data-anchor-id="log-likelihood-ratio-curve">(log) Likelihood Ratio Curve</h3>
<p>The <code>ppv</code> depends on the empirical densities of the biomarkers only through their ratio. This ratio corresponds to the likelihood ratio when using Bayes’ rule. Higher values of the ZNF487 biomarker results in a higher ratio which in turn indicates a higher probability of optimal debulking. The next plot is the empirical log-likelihood ratio of the optimal vs.&nbsp;suboptimal debulking outcomes, as a function of the gene expression level.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ff0 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>ff1 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ffx <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>x</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ffx,<span class="fu">log</span>(ff1<span class="sc">/</span>ff0),<span class="at">xlab=</span><span class="fu">paste</span>(<span class="st">"Log Expression"</span>,GeneName),<span class="at">ylab=</span><span class="st">"Log Density Ratio"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Log-likelihood Ratio for ZNF487 expression</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="threshold-based-classification" class="level2">
<h2 class="anchored" data-anchor-id="threshold-based-classification">Threshold-based Classification</h2>
<section id="x2-tables" class="level3">
<h3 class="anchored" data-anchor-id="x2-tables">2x2 tables</h3>
<p>For this discussion, a classifier is a hard and fast decision rule to separate patients in two groups for the purpose of treatment decisions. The simplest classifier we can construct using this biomarker is to establish a threshold such that patients with expression above the threshold are classified as likely optimal. In keeping with medical terminology we refer to the individuals with the more severe disease, in this case the suboptimally debulked patients, as “positive”. If we were to decide to set threshold at <span class="math inline">\(-1\)</span>, the reported confusion matrix associated with such a decision rule would be given by the table below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(XX <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">1</span>,YY<span class="sc">==</span><span class="dv">0</span>,<span class="at">dnn=</span><span class="fu">list</span>(<span class="st">"Biomarker Positive"</span>,<span class="st">"True Positive"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                  True Positive
Biomarker Positive FALSE TRUE
             FALSE     7    0
             TRUE     12   21</code></pre>
</div>
</div>
<p>We observe that all true suboptimal debulking cases were correctly classified. We say that the true positive rate (or sensitivity) is <span class="math inline">\(1\)</span>. On the other hand among the true optimal debulking cases, <span class="math inline">\(7\)</span> out of <span class="math inline">\(19\)</span> were correctly predicted. The true negative rate is then <span class="math inline">\(7/19\)</span> corresponding to a specificity of <span class="math inline">\(1 - 7/19\)</span>.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Discussion Point
</div>
</div>
<div class="callout-body-container callout-body">
<p>In which scenarios is is this a good cutoff?</p>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="visualizing-properties-of-alternative-thresholds" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-properties-of-alternative-thresholds">Visualizing Properties of Alternative Thresholds</h3>
<p>There are many valid alternatives for investigating and visualizing the properties of a threshold-based classifier. These include</p>
<ul>
<li><p>Receiver Operating Characteristic (ROC) curves, which plots sensitivity against one minus specificity</p></li>
<li><p>Precision versus Recall plots, which plots positive predictive value against sensitivity</p></li>
<li><p>Total Operating Characteristic. <a href="https://doi.org/10.1080/13658816.2013.862623">@pontius2014</a></p></li>
<li><p>Detection Error Tradeoff <a href="https://ccc.inaoep.mx/~villasen/bib/martin97det.pdf">@Martin:1997ve</a>.</p></li>
</ul>
<p><a href="https://doi.org/10.1371/journal.pone.0118432">@saito2015</a> compares ROC and PR in some detail.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="the-roc-curve" class="level3">
<h3 class="anchored" data-anchor-id="the-roc-curve">The ROC curve</h3>
<p>We will now motivate and introduce the concept of the ROC Curve. It is useful to begin with cumulative distributions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">ecdf</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>]),<span class="at">main=</span><span class="st">""</span>,<span class="at">ylab=</span><span class="st">"Empirical CDFs"</span>,<span class="at">col=</span><span class="st">"blue"</span>,<span class="at">xlim=</span><span class="fu">range</span>(XX))</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">ecdf</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>]),<span class="at">col=</span><span class="st">"green"</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v=</span><span class="sc">-</span><span class="dv">1</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Empirical CDFs of ZNF487 for Optimal and Suboptimal debulking</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>In the figure above, the blue curve represents the empirical CDFs of the suboptimal debulking biomarker values, and the green curve the empirical CDFs of the optimal. One can then imagine setting many thresholds across the biomarker expression values (x-axis) and pairing the blue/green CDF values obtained. For example the earlier threshold of <span class="math inline">\(-1\)</span> is also shown. It crosses the blue line at the true positive rate, and the green line at the false negative rate, or one minus the true negative rate.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Type 'citation("pROC")' for a citation.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'pROC'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:BiocGenerics':

    var</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    cov, smooth, var</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">roc</span>(YY,XX),<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Empirical AUC of ZNF487 Expression for discrimination of Optimal v. Suboptimal</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>The empirical ROC curve can then be computed by comparing the pairings from many such thresholds. The ROC curve is a 2-D plot with the true positive rate (sensitivity) on the y-axis and the false positive rate (one minus specificity) on the x-axis. It yields the tradeoffs between these two rates for a given cutoff. In other words, each point on the curve represents the true and false positive rates associated with a threshold. The 45 degree line (colored red) of the ROC curve represents the curve one would obtain if the two cumulative distributions were the same. The ROC curve can be used to informally assess the relative merits of alternative cutoffs, as well as the overall separation of the curves. Directly leading from the ROC Curve is the AUC (Area under the Curve). This is a widely used metric for comparing different classification methods. The AUC is represented visually above as the area under the black curve (including the area to the right of the red line).</p>
<div style="page-break-after: always;"></div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The general rule of thumb is that given a classification method and two biomarkers, the biomarker with the higher AUC under the same method tends to be a better classifier.</p>
<p>What are some limitations of this metric?</p>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>We can also build a smooth version of the ROC curve, for example by calculating estimates of the class-conditional cumulative distributions from the smooth density plot we derived earlier.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>FF0 <span class="ot">=</span> <span class="fu">cumsum</span>(ff0); FF0 <span class="ot">=</span> FF0 <span class="sc">/</span> <span class="fu">max</span>(FF0)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>FF1 <span class="ot">=</span> <span class="fu">cumsum</span>(ff1); FF1 <span class="ot">=</span> FF1 <span class="sc">/</span> <span class="fu">max</span>(FF1)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">-</span>FF0,<span class="dv">1</span><span class="sc">-</span>FF1,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"False Positive Rate"</span>,GeneName),</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"True Positive Rate"</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">ROC for ZNF487 derived from class-conditional smooth density estimates</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>The AUC nonparametrically compares the two distributions, and it turns out that testing for whether the ROC significantly deviates from the diagonal line can be done equivalently by the non-parametric rank-sum test, which is implemented below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>],XX[YY<span class="sc">==</span><span class="cn">FALSE</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  XX[YY == TRUE] and XX[YY == FALSE]
W = 340, p-value = 6.261e-05
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>The null hypothesis is that the biomarker level of a randomly selected sample from the set of optimal cases will have the same chance of being greater than or less than a randomly selected sample from the set of suboptimal cases. The p-value above strongly indicates potential departure from the null.</p>
<div style="page-break-after: always;"></div>
<p>When searching for biomarkers, a statistically significant difference from the diagonal may or may not be associated with a clinical useful biomarker. For example, let us examine the dot-plots of the ANKRD66 gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>GeneName <span class="ot">=</span> <span class="st">"ANKRD66"</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>XX_ANKRD66 <span class="ot">=</span> <span class="fu">exprs</span>(GSE32063_eset)[GeneName,]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>YY_ANKRD66 <span class="ot">=</span> <span class="fu">pData</span>(GSE32063_eset)[,<span class="st">"debulking"</span>] <span class="sc">==</span> <span class="st">"optimal"</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(XX_ANKRD66,YY_ANKRD66,</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression of"</span>,GeneName),<span class="at">cex=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Dotplot for ANKRD66</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
<p>The dot-plot gives no visual indication that there is good separation of biomarker expression across the two debulking outcomes. However, the p-value from the rank sum test below for these two outcomes is <span class="math inline">\(0.05042\)</span>, which, although not technically significant at the traditional <span class="math inline">\(.05\)</span> threshold, provides at least some indication of a departure from equal distributions. This departure is primarily attributable to the a) two outliers in the suboptimal class and b) an accumulation of points slightly above -4 in the optimal class which has a strong effect on the ranks but happens within a range unlikely to translate into biologically meaningful activity. Two points emerge: differences in ranks do not necessarily translate in clinically useful biomarkers. Lack of overlap in the tails can quickly produce significant Wilcoxon tests.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">wilcox.test</span>(XX_ANKRD66[YY_ANKRD66<span class="sc">==</span><span class="cn">TRUE</span>],</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>            XX_ANKRD66[YY_ANKRD66<span class="sc">==</span><span class="cn">FALSE</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum exact test

data:  XX_ANKRD66[YY_ANKRD66 == TRUE] and XX_ANKRD66[YY_ANKRD66 == FALSE]
W = 272, p-value = 0.05042
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>An even more important disconnect between statistical significance and clinical utility occurs in larger studies, where biomarker may achieve a rejection of the null as a result of biologically weak differences between the two class conditional distributions.</p>
<div style="page-break-after: always;"></div>
<p>Finally, we discuss on how the ROC relates to the CDF and the labeling of positive and negative cases. Suppose that large or high values of the biomarker are associated with positive cases. In such a case, the ROC curve is created by plotting the following function:</p>
<p><span class="math display">\[
ROC_p = \{1-F_0(c), 1-F_1(c)\} \qquad \text{for} \ -\infty &lt; c &lt; \infty
\]</span></p>
<p>where <span class="math inline">\(F_1(x), F_0(x)\)</span> are the CDFs of the positive and negative cases, respectively. The ROC plot in such a case is creating by plotting the above pairing across many thresholds <span class="math inline">\(c\)</span>. If instead it is the case that large or high values of the biomarker are associated with negative cases, then the ROC curve is created by plotting instead:</p>
<p><span class="math display">\[
ROC_n = \{F_0(c), F_1(c)\} \qquad \text{for} \ -\infty &lt; c &lt; \infty
\]</span></p>
<p>We will adopt this convention to avoid any labeling confusions.</p>
<p>A fun fact is that the vertical distance between the diagonal and the ROC curve adds to Sensitivity + Specificity -1, a quantity called the Youden’s index and sometime used to pick a threshold.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="the-precision-recall-curve" class="level3">
<h3 class="anchored" data-anchor-id="the-precision-recall-curve">The Precision / Recall curve</h3>
<p>The logic of the Precision recall curve is similar: vary the cutoff and graph two interesting summary of biomarker performance. In this case the summaries are sensitivity and PPV. I find this closer to the relevant clinical question. The formula is</p>
<p><span class="math display">\[
PR_p = \{ 1-F_1(c), \pi F_1(c) / [\pi F_1(c) + (1-\pi)F_0(c)] \} \qquad \text{for} \ -\infty &lt; c &lt; \infty
\]</span> By default, most packages will create PR curves directly from the empirical 2x2 tables generated at various cutoffs, which means that in this formula the prevalence <span class="math inline">\(\pi\)</span> is set at the study proportion of positive cases. What does this tells us about comparing PR curves across studies?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>roc_ZNF487 <span class="ot">=</span> <span class="fu">roc</span>(YY,XX)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(precision <span class="sc">~</span> recall,</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coords</span>(roc_ZNF487, <span class="st">"all"</span>, <span class="at">ret =</span> <span class="fu">c</span>(<span class="st">"recall"</span>, <span class="st">"precision"</span>), <span class="at">transpose =</span> <span class="cn">FALSE</span>),</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">type=</span><span class="st">"l"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Empirical AUC of ZNF487 Expression for discrimination of Optimal v. Suboptimal</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="the-det-curve" class="level3">
<h3 class="anchored" data-anchor-id="the-det-curve">The DET curve</h3>
<p>For a final digression, the DET curve also visualizes properties of the biomarker by summarizing error rates at many cutoffs. Unlike the ROC, it considers the false negative rate and false positive rates. Rather than starting with a grid of possible cutoffs, one starts with a grid of possible error rates. In one implementation, you pick the same error rate for both false alarms and false negatives. Then you compute the quantiles corresponding to those error rates in the biomarker scale, and standardize them by subtracting the class-conditional mean and dividing by the class-conditional standard deviation. The DET plots the results against each other. If the distributions are the same, the graph is the <span class="math inline">\(Y=-X\)</span> line. Good biomarkers generate lines that lie below.</p>
<p>I have not seen much use. It seems cool but it does take a bit to get used to it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>DET <span class="ot">=</span> <span class="cf">function</span>(XX,YY){</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>   mm0 <span class="ot">=</span> <span class="fu">mean</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>])</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>   sd0 <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="fu">var</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>]))</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>   mm1 <span class="ot">=</span> <span class="fu">mean</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>])</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>   sd1 <span class="ot">=</span> <span class="fu">sqrt</span>(<span class="fu">var</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>]))</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>   pp <span class="ot">=</span> <span class="fu">seq</span>(<span class="fl">0.01</span>,.<span class="dv">99</span>,<span class="at">length.out=</span><span class="dv">128</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>   zz0 <span class="ot">=</span> ( <span class="fu">quantile</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>], <span class="at">probs =</span> <span class="dv">1</span><span class="sc">-</span>pp) <span class="sc">-</span> mm0 ) <span class="sc">/</span> sd0</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>   zz1 <span class="ot">=</span> ( <span class="fu">quantile</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>], <span class="at">probs =</span> pp) <span class="sc">-</span> mm1 ) <span class="sc">/</span> sd1</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>   output <span class="ot">=</span> <span class="fu">list</span>(<span class="at">zz0=</span>zz0,<span class="at">zz1=</span>zz1)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>   <span class="fu">return</span>(output)</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>DET_ZNF487 <span class="ot">=</span> <span class="fu">DET</span>(XX,YY)</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(DET_ZNF487<span class="sc">$</span>zz0,DET_ZNF487<span class="sc">$</span>zz1,</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"False Positive Rate Deviate"</span>,GeneName),</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"False Negative Rate Deviate"</span>, <span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">1</span>,<span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">DET curve for ZNF487 derived from class-conditional smooth density estimates</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="comparing-znf487-and-igha1" class="level2">
<h2 class="anchored" data-anchor-id="comparing-znf487-and-igha1">Comparing ZNF487 and IGHA1</h2>
<p>IGHA1 is another gene that behaves differently in the optimal vs.&nbsp;suboptimal group. The pattern is a change in both location and scale. The shift is large but there is no “clean” region where only one group is present. Also the directionality is reversed compared to ZNF487 so to make things comparable we will change the sign.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>GeneName <span class="ot">=</span> <span class="st">"ZNF585B"</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>XX_ZNF585B <span class="ot">=</span> <span class="fu">exprs</span>(GSE32063_eset)[GeneName,]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>YY <span class="ot">=</span> <span class="fu">pData</span>(GSE32063_eset)[,<span class="st">"debulking"</span>] <span class="sc">==</span> <span class="st">"optimal"</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(XX_ZNF585B,YY,<span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression of"</span>,GeneName),</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Debulking Status"</span>,<span class="at">cex=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">Dotplot for ZNF585B</figcaption>
</figure>
</div>
</div>
</div>
<section id="ppv-comparison" class="level3">
<h3 class="anchored" data-anchor-id="ppv-comparison">PPV comparison</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>ffx <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>x</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>ff0 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>ff1 <span class="ot">=</span> <span class="fu">density</span>(XX[YY<span class="sc">==</span><span class="cn">TRUE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>prev <span class="ot">=</span> .<span class="dv">2</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>ppv_ZNF487 <span class="ot">=</span> prev <span class="sc">*</span> ff1 <span class="sc">/</span> ( prev <span class="sc">*</span> ff1 <span class="sc">+</span> (<span class="dv">1</span><span class="sc">-</span>prev) <span class="sc">*</span> ff0)</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ffx,ppv_ZNF487,<span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),<span class="at">col=</span><span class="st">"blue"</span>,</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"Log Expression"</span>,GeneName),</span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"Probability of optimal debulking"</span>,<span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>ffx <span class="ot">=</span> <span class="fu">density</span>(XX_ZNF585B[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>x</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>ff0 <span class="ot">=</span> <span class="fu">density</span>(XX_ZNF585B[YY<span class="sc">==</span><span class="cn">FALSE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>ff1 <span class="ot">=</span> <span class="fu">density</span>(XX_ZNF585B[YY<span class="sc">==</span><span class="cn">TRUE</span>],<span class="at">from=</span><span class="sc">-</span><span class="dv">4</span>,<span class="at">to=</span><span class="dv">0</span>)<span class="sc">$</span>y</span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>ppv_ZNF585B <span class="ot">=</span> prev <span class="sc">*</span> ff1 <span class="sc">/</span> ( prev <span class="sc">*</span> ff1 <span class="sc">+</span> (<span class="dv">1</span><span class="sc">-</span>prev) <span class="sc">*</span> ff0)</span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(ffx,ppv_ZNF585B,<span class="at">col=</span><span class="st">"orange"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="960"></p>
<figcaption class="figure-caption">PPV Plot for ZNF487 and IGHA1 expression</figcaption>
</figure>
</div>
</div>
</div>
<p>Open for comments.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="roc-comparison" class="level3">
<h3 class="anchored" data-anchor-id="roc-comparison">ROC comparison</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pROC)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">roc</span>(YY,XX),<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">roc</span>(YY,XX_ZNF585B),<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">"orange"</span>,<span class="at">add=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE
Setting direction: controls &lt; cases</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Empirical AUC of ZNF487 Expression for discrimination of Optimal v. Suboptimal</figcaption>
</figure>
</div>
</div>
</div>
<p>Open for comments.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="pr-comparison" class="level3">
<h3 class="anchored" data-anchor-id="pr-comparison">PR comparison</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(precision <span class="sc">~</span> recall,</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coords</span>(<span class="fu">roc</span>(YY,XX), <span class="st">"all"</span>, <span class="at">ret =</span> <span class="fu">c</span>(<span class="st">"recall"</span>, <span class="st">"precision"</span>), <span class="at">transpose =</span> <span class="cn">FALSE</span>),</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">type=</span><span class="st">"l"</span>,<span class="at">col=</span><span class="st">"blue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(precision <span class="sc">~</span> recall,</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>     <span class="fu">coords</span>(<span class="fu">roc</span>(YY,XX_ZNF585B), <span class="st">"all"</span>, <span class="at">ret =</span> <span class="fu">c</span>(<span class="st">"recall"</span>, <span class="st">"precision"</span>), <span class="at">transpose =</span> <span class="cn">FALSE</span>),</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">type=</span><span class="st">"l"</span>,<span class="at">col=</span><span class="st">"orange"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE
Setting direction: controls &lt; cases</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting levels: control = FALSE, case = TRUE</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Setting direction: controls &lt; cases</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Empirical PR Curve of ZNF487 Expression for discrimination of Optimal v. Suboptimal</figcaption>
</figure>
</div>
</div>
</div>
<p>Open for comments.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="det-comparison" class="level3">
<h3 class="anchored" data-anchor-id="det-comparison">DET comparison</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>DET_ZNF585B <span class="ot">=</span> <span class="fu">DET</span>(XX_ZNF585B,YY)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(DET_ZNF487<span class="sc">$</span>zz0,DET_ZNF487<span class="sc">$</span>zz1,</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="fu">c</span>(<span class="st">"False Positive Rate Deviate"</span>,GeneName),</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab=</span><span class="st">"False Negative Rate Deviate"</span>, <span class="at">type =</span> <span class="st">"l"</span>,<span class="at">lwd=</span><span class="dv">2</span>,<span class="at">col=</span><span class="st">"blue"</span>)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(DET_ZNF585B<span class="sc">$</span>zz0,DET_ZNF585B<span class="sc">$</span>zz1,<span class="at">col=</span><span class="st">"orange"</span>,<span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">1</span>,<span class="sc">-</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L4-metrics_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">DET curve for ZNF487 derived from class-conditional smooth density estimates</figcaption>
</figure>
</div>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="the-skill-plot" class="level3">
<h3 class="anchored" data-anchor-id="the-skill-plot">The Skill Plot</h3>
<p>[https://doi.org/10.1111/j.1541-0420.2007.00781_1.x]</p>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>
<div id="refs" role="list">

</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>