<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.545">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>STAT 117: Data Analysis in Modern Biostatistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">STAT 117: Data Analysis in Modern Biostatistics</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="./index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="./about.html"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-lectures" role="button" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Lectures</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-lectures">    
        <li>
    <a class="dropdown-item" href="./L1-intro.html">
 <span class="dropdown-text">Lecture 1 Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L2-biomarkers.html">
 <span class="dropdown-text">Lecture 2: Cancer Biomarkers</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L3-data.html">
 <span class="dropdown-text">Lecture 3: Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L4-metrics.html">
 <span class="dropdown-text">Lecture 4: Basics of Biomarker Evaluation</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L5-decision.html">
 <span class="dropdown-text">Lecture 5: Decision Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L6-bayes.html">
 <span class="dropdown-text">Lecture 6: Bayesian Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L7-mcmc.html">
 <span class="dropdown-text">Lecture 7: Monte Carlo Markov Chains</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L8-timetoevent.html">
 <span class="dropdown-text">Lecture 8: Time to Event Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L9-comparing.html">
 <span class="dropdown-text">Lecture 9: Comparing Criteria for Biomarker Discovery</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L10-fdr.html">
 <span class="dropdown-text">Lecture 10: False Discovery Rates</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L11-regmean.html">
 <span class="dropdown-text">Lecture 11: Regression to the Mean</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L12-shrinkage.html">
 <span class="dropdown-text">Lecture 12: Shrinkage</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L14-shrinkage2.html">
 <span class="dropdown-text">Lecture 14: Shrinkage part 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L15-metaanalysis2.html">
 <span class="dropdown-text">Lecture 15: Metaanalysis part 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./L17-twomarkers.html">
 <span class="dropdown-text">Lecture 17</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#lecture-10-false-discovery-rates" id="toc-lecture-10-false-discovery-rates" class="nav-link active" data-scroll-target="#lecture-10-false-discovery-rates">Lecture 10: False Discovery Rates</a>
  <ul class="collapse">
  <li><a href="#multiple-testing" id="toc-multiple-testing" class="nav-link" data-scroll-target="#multiple-testing">Multiple Testing</a>
  <ul class="collapse">
  <li><a href="#intro" id="toc-intro" class="nav-link" data-scroll-target="#intro">Intro</a></li>
  <li><a href="#setting-and-notation" id="toc-setting-and-notation" class="nav-link" data-scroll-target="#setting-and-notation">Setting and notation:</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#distribution-of-p-values" id="toc-distribution-of-p-values" class="nav-link" data-scroll-target="#distribution-of-p-values">Distribution of p-values</a></li>
  </ul></li>
  <li><a href="#false-discovery-rates" id="toc-false-discovery-rates" class="nav-link" data-scroll-target="#false-discovery-rates">False Discovery Rates</a>
  <ul class="collapse">
  <li><a href="#frequentist-definition-of-fdr" id="toc-frequentist-definition-of-fdr" class="nav-link" data-scroll-target="#frequentist-definition-of-fdr">Frequentist Definition of FDR</a></li>
  <li><a href="#the-benjamini-hochberg-algorithm" id="toc-the-benjamini-hochberg-algorithm" class="nav-link" data-scroll-target="#the-benjamini-hochberg-algorithm">The Benjamini-Hochberg Algorithm</a></li>
  <li><a href="#empirical-null-distribtution-of-scores" id="toc-empirical-null-distribtution-of-scores" class="nav-link" data-scroll-target="#empirical-null-distribtution-of-scores">Empirical Null Distribtution of Scores</a></li>
  <li><a href="#empirical-bayes-estimate-of-the-false-discovery-rate" id="toc-empirical-bayes-estimate-of-the-false-discovery-rate" class="nav-link" data-scroll-target="#empirical-bayes-estimate-of-the-false-discovery-rate">Empirical Bayes estimate of the false discovery rate:</a></li>
  <li><a href="#empirical-bayes-estimate-of-the-local-fdr" id="toc-empirical-bayes-estimate-of-the-local-fdr" class="nav-link" data-scroll-target="#empirical-bayes-estimate-of-the-local-fdr">Empirical Bayes estimate of the local FDR</a></li>
  <li><a href="#addendum-familywise-error-control" id="toc-addendum-familywise-error-control" class="nav-link" data-scroll-target="#addendum-familywise-error-control">Addendum: Familywise Error Control</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">




<div style="page-break-after: always;"></div>
<section id="lecture-10-false-discovery-rates" class="level1">
<h1>Lecture 10: False Discovery Rates</h1>
<section id="multiple-testing" class="level2">
<h2 class="anchored" data-anchor-id="multiple-testing">Multiple Testing</h2>
<section id="intro" class="level3">
<h3 class="anchored" data-anchor-id="intro">Intro</h3>
<p>In our previous discussion we played with various approaches for discovering sets of promising biomarkers. How can we assess the quality of the discovery process and results? The most basic version of this assessment is based on assuming there is a binary truth at the gene level, representing biomarker with and without biological signal, and that the discovery process is about revealing this vector of binary labels. Formally, this is a multiple hypothesis testing problem.</p>
<p>The literature of multiple testing is enormously big. A favorite of mine is <a href="https://doi.org/10.1214/ss/1177011945">@Tukey:1991ga</a></p>
</section>
<section id="setting-and-notation" class="level3">
<h3 class="anchored" data-anchor-id="setting-and-notation">Setting and notation:</h3>
<p><span class="math inline">\(g = 1, \ldots, G\)</span> genomic features, for example genes in gene-level analysis of expression data</p>
<p><span class="math inline">\(i = 1, \ldots, I\)</span> subjects in the discovery sample</p>
<p><span class="math inline">\(x_{1g}, \ldots, x_{Ig}\)</span> observed quantitative level of the expression of gene <span class="math inline">\(g\)</span> in the <span class="math inline">\(I\)</span> subjects</p>
<p><span class="math inline">\(y_{1}, \ldots, y_{I}\)</span> binary class labels for the <span class="math inline">\(I\)</span> subjects</p>
<p><span class="math inline">\(z_g = z_g (x_{1g}, \ldots, x_{Ig}, y_{1}, \ldots, y_{I})\)</span> a statistic used to select, or “discover” biomarkers. For example the absolute value of a t-statistic, the AUC or the fold change. We assume that larger values of the statistic are indicative of the signal we try to discover.</p>
<p><span class="math inline">\(H_{0g}\)</span> is the event that for gene <span class="math inline">\(g\)</span> there is no signal, that is if the two classes have identical distributions for the expression <span class="math inline">\(x\)</span>.</p>
<p><span class="math inline">\(p_g = P ( z_g &gt; c | H_{0g})\)</span> p-value.</p>
<div style="page-break-after: always;"></div>
</section>
<section id="data" class="level3">
<h3 class="anchored" data-anchor-id="data">Data</h3>
<p>Continuous genomic feature: gene expression microarray readout in the TCGA study</p>
<p>Binary phenotype (optimal surgical debulking)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(curatedOvarianData)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ROCR)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(TCGA_eset)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>XX <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">cbind</span>(<span class="fu">exprs</span>(TCGA_eset)))</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>YY <span class="ot">=</span> <span class="dv">1</span> <span class="sc">*</span> <span class="fu">as.vector</span>(<span class="fu">pData</span>(TCGA_eset)[,<span class="st">"debulking"</span>]<span class="sc">==</span><span class="st">"optimal"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>XX <span class="ot">=</span> XX[,<span class="sc">!</span><span class="fu">is.na</span>(YY)]; </span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>YY <span class="ot">=</span> YY[<span class="sc">!</span><span class="fu">is.na</span>(YY)]</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>XXX <span class="ot">=</span> XX</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>YYY <span class="ot">=</span> YY</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># subset</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>subs <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>XX <span class="ot">=</span> XX[,subs]; YY <span class="ot">=</span> YY[subs]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="distribution-of-p-values" class="level3">
<h3 class="anchored" data-anchor-id="distribution-of-p-values">Distribution of p-values</h3>
<p>We next create two sets of summary scores. <code>ScoresSub</code> yields the four metrics computed on the first 100 patients, while <code>ScoresAll</code> gives the four metrics computed on the entire set of patients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"RScripts/Scores.R"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>ScoresSub <span class="ot">=</span> <span class="fu">CompScores</span>(XX,YY)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>ScoresAll <span class="ot">=</span> <span class="fu">CompScores</span>(XXX,YYY)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A useful place to start a discussion about multiple testing are the histograms of the p-values obtained with the t-test.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">pty =</span> <span class="st">"s"</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">exp</span>(<span class="sc">-</span>ScoresSub[,<span class="st">"nlpvalueT"</span>]),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Subset"</span>,<span class="at">xlab=</span><span class="st">"p-value"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">exp</span>(<span class="sc">-</span>ScoresAll[,<span class="st">"nlpvalueT"</span>]),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Full Set"</span>,<span class="at">xlab=</span><span class="st">"p-value"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L10-fdr_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Histograms of P-values, for the 100 patient subset (left) and Entire Set</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">pty =</span> <span class="st">"s"</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">ecdf</span>(<span class="fu">exp</span>(<span class="sc">-</span>ScoresSub[,<span class="st">"nlpvalueT"</span>])),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Subset"</span>,<span class="at">xlab=</span><span class="st">"p-value"</span>,<span class="at">ylab=</span><span class="st">"Empirical CDF"</span>); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">ecdf</span>(<span class="fu">exp</span>(<span class="sc">-</span>ScoresAll[,<span class="st">"nlpvalueT"</span>])),</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Full Set"</span>,<span class="at">xlab=</span><span class="st">"p-value"</span>,<span class="at">ylab=</span><span class="st">"Empirical CDF"</span> ); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L10-fdr_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Empirical CDFs of P-values, for the 100 patient subset (left) and Entire Set</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Higher Criticism
</div>
</div>
<div class="callout-body-container callout-body">
<p>Review: if the null hypothesis is correct, the sampling distribution of the p-value is uniform in (0,1). “Sampling distribution” refers to hypothetical repetitions of the same experiment, in this case under the null.</p>
<ul>
<li><p>These histograms show a single experiment. The distribution arises from variation across biomarkers. Can we still learn something useful about the frequentist multiple testing problem?</p></li>
<li><p>How would you test for the “global null”, that is the scenario where each and every biomarker is independent of the debulking status?</p></li>
</ul>
</div>
</div>
<p>Hint: <a href="http://projecteuclid.org/euclid.ss/1425492437">@Donoho2015ss</a></p>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="false-discovery-rates" class="level2">
<h2 class="anchored" data-anchor-id="false-discovery-rates">False Discovery Rates</h2>
<p>A simple way to think about the discovery process is to focus on a list of promising candidates (or “discoveries”), as we did in our previous lecture. If we know the true data generating model for each gene, we could, for example, look at this table:</p>
<p>It is common to study the proportion <span class="math inline">\(R_0/R\)</span> of discoveries for which the true data generating model is null.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><span class="math inline">\(R_0/R\)</span> is unknown. It depends on both the data and the parameter (the vector of indicators of whether eagh gene is null)</p>
<ul>
<li>When is <span class="math inline">\(R_0/R\)</span> an appropriate quantity to focus on?</li>
</ul>
</div>
</div>
<p>This parameter can be estimated from a couple of different perspectives. If you are interested in evaluating the list generating <em>process</em> you may be interested in a frequentist expectation. Bounds can be obtained using methods like Benjamini-Hochberg (see below). If your focus is on the actual list rather than the procedure, Bayesian estimates may be your choice.</p>
<p>We review many of the connections here: <a href="https://biostats.bepress.com/jhubiostat/paper115/">@muel:parm:rice:2007</a></p>
<div style="page-break-after: always;"></div>
<section id="frequentist-definition-of-fdr" class="level3">
<h3 class="anchored" data-anchor-id="frequentist-definition-of-fdr">Frequentist Definition of FDR</h3>
<div style="page-break-after: always;"></div>
</section>
<section id="the-benjamini-hochberg-algorithm" class="level3">
<h3 class="anchored" data-anchor-id="the-benjamini-hochberg-algorithm">The Benjamini-Hochberg Algorithm</h3>
<p></p>
<div style="page-break-after: always;"></div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>rankedPValuesSub <span class="ot">=</span> <span class="fu">sort</span>( <span class="fu">exp</span>(<span class="sc">-</span>ScoresSub[,<span class="st">"nlpvalueT"</span>]) )</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>rankedPValuesAll <span class="ot">=</span> <span class="fu">sort</span>( <span class="fu">exp</span>(<span class="sc">-</span>ScoresAll[,<span class="st">"nlpvalueT"</span>]) )</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">pty =</span> <span class="st">"s"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( (<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)<span class="sc">/</span><span class="fu">length</span>(rankedPValuesSub), rankedPValuesSub[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>],</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">"Rank/NGenes"</span>, <span class="at">ylab=</span><span class="st">"p-value"</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,.<span class="dv">07</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>( (<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)<span class="sc">/</span><span class="fu">length</span>(rankedPValuesAll), rankedPValuesAll[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>],</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>      <span class="at">xlab=</span><span class="st">"Rank/NGenes"</span>, <span class="at">ylab=</span><span class="st">"p-value"</span> )</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="dv">0</span>,.<span class="dv">07</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="L10-fdr_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Benjamini-Hochberg Plots</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>How would you comment these results?</p>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="empirical-null-distribtution-of-scores" class="level3">
<h3 class="anchored" data-anchor-id="empirical-null-distribtution-of-scores">Empirical Null Distribtution of Scores</h3>
<p>It is simple to permute the labels of the subjects and recompute the scores. After permutation, for each gene, the link between expression and debulking has been broken. Discoveries are now random. We will look at a single permutation for illustration. As the number of features is large a single permutation can give reasonable indications. In a real application we may want to use multiple permutation and average over the inferential results, particularly if we are interested in events with small probability.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>YYNull <span class="ot">=</span> YY[ <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(YY)) ]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>ScoresSubNull <span class="ot">=</span> <span class="fu">CompScores</span>(XX,YYNull)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>YYYNull <span class="ot">=</span> YYY[ <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(YYY)) ]</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>ScoresAllNull <span class="ot">=</span> <span class="fu">CompScores</span>(XXX,YYYNull)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>acut <span class="ot">=</span> .<span class="dv">6</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>aucDiscAll <span class="ot">=</span> ScoresAll[,<span class="st">"AUC"</span>]<span class="sc">&gt;</span>acut</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>aucDiscAllNull <span class="ot">=</span> ScoresAllNull[,<span class="st">"AUC"</span>]<span class="sc">&gt;</span>acut</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(aucDiscAll)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>aucDiscAll
FALSE  TRUE 
13076    28 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(aucDiscAllNull)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>aucDiscAllNull
FALSE  TRUE 
13103     1 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p>is it correct to conclude that of the 28 discovery, only 1 is likely false?</p></li>
<li><p>what is 1/28 and estimate of?</p></li>
</ul>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="empirical-bayes-estimate-of-the-false-discovery-rate" class="level3">
<h3 class="anchored" data-anchor-id="empirical-bayes-estimate-of-the-false-discovery-rate">Empirical Bayes estimate of the false discovery rate:</h3>
<p>Empirical Bayes FDR makes the 1/28 story a bit more formal. The thinking is now conditional on the observed set of data, specifically the observed set of scores, the <span class="math inline">\(z\)</span>’s. The <span class="math inline">\(z\)</span>’s can coincide with the p-values, but this logic does not require p-values, so the p-value calculation adds an unnecessary step.</p>
<p>Define <span class="math inline">\(F(z) = P ( z_g \leq z)\)</span> to the the true marginal distribution of the scores, and <span class="math inline">\(\bar F(z) = 1 - F(z) = P ( z_g &gt; z)\)</span>. Now restrict attention to the genes whose true generating model is null and let <span class="math inline">\(\bar F_0(z) = 1 - F_0(z) = P ( z_g &gt; z | H_{og})\)</span>. In a genemo where <span class="math inline">\(p_A\)</span> genes are from some alternative (each gene can have its own!) and the remaining <span class="math inline">\(1-p_A\)</span> are from the same null, the unknown proportion of false discoveries can be rewritten as:</p>
<p><span class="math display">\[
\frac{R_0}{R} = \frac{ (1-p_A) * \bar F_0 (z)}{(1-p_A) * \bar F_0 (z) + p_A \bar F_A (z) } =\frac{  (1-p_A) \bar F_0 (z)}{ \bar F (z) } \approx\frac{ \bar F_0 (z)}{ \bar F (z) }
\]</span></p>
<p>In EB, the denominator, for any given cutoff, can be estimated empirically from the geneome-wide distribution of <span class="math inline">\(z_g\)</span>’s. Just don’t pick the cutoff so that the list is empty.</p>
<p>The numerator is a bit trickier. If the disctribution of the <span class="math inline">\(z\)</span>’s under the null is known (as is the case for example with <span class="math inline">\(z\)</span>-scores for which normality can be trusted, <span class="math inline">\(\bar F_0 (z)\)</span> can be estimated that way. If not, one approach is to use the geneome-wide distribution of <span class="math inline">\(z_g\)</span>’s after a permutation that mimics the global null. <span class="math inline">\(p_A\)</span> is not identifiable w/out parametric / smoothness assumptions on the distributions of the <span class="math inline">\(z\)</span>’s. A surprisingly useful approximation is <span class="math inline">\(1-p_A = 1\)</span>, which gives a conservative bounds to EF FDR, but often a reasonably realistic one.</p>
<p><a href="https://doi.org/10.1214/aos/1051027871">@Efron2003as</a></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>tie this back to the 1/28 story</li>
</ul>
</div>
</div>
<div style="page-break-after: always;"></div>
</section>
<section id="empirical-bayes-estimate-of-the-local-fdr" class="level3">
<h3 class="anchored" data-anchor-id="empirical-bayes-estimate-of-the-local-fdr">Empirical Bayes estimate of the local FDR</h3>
<p>Let’s look at estimates of densities corresponding to <span class="math inline">\(F\)</span> and <span class="math inline">\(F_0\)</span>. These densities are normalized to integrate to <span class="math inline">\(1\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">density</span>(ScoresAllNull[,<span class="st">"nlpvalueT"</span>]),<span class="at">col=</span><span class="fu">gray</span>(.<span class="dv">7</span>),<span class="at">lwd=</span><span class="dv">4</span>,<span class="at">xlim=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">6</span>),</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">""</span>,<span class="at">xlab=</span><span class="st">"z score"</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(<span class="fu">density</span>(ScoresAll[,<span class="st">"nlpvalueT"</span>]),<span class="at">col=</span><span class="dv">2</span>,<span class="at">lwd=</span><span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="L10-fdr_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>does this figure suggests a quick and dirty way of estimating <span class="math inline">\(1-p_A\)</span>? Think through the assumptions would would be making.</li>
</ul>
</div>
</div>
<p>This figure does suggest a way of evaluating the false discovery rate specifically for discoveries at and around a score of <span class="math inline">\(z\)</span>. Imagine you have estimates <span class="math inline">\(\hat f\)</span> and <span class="math inline">\(\hat f_0\)</span>.</p>
<p><span class="math display">\[
\text{fdr}(z) = \frac{ (1-\hat p_A) * \hat f_0 (z)}{(1-\hat p_A) * \hat f_0 (z) + \hat p_A f_A (z) } =\frac{  (1-\hat p_A) \hat f_0 (z)}{ \hat f (z) } \approx\frac{ \hat f_0 (z)}{ \hat f (z) }
\]</span> ### Summary</p>
<p>If you can trust the assumptions behind the p-value calculation, then:</p>
<p>else:</p>
<div style="page-break-after: always;"></div>
</section>
<section id="addendum-familywise-error-control" class="level3">
<h3 class="anchored" data-anchor-id="addendum-familywise-error-control">Addendum: Familywise Error Control</h3>
<p>A different way of thinking about multiple testing is to control the probability of falsely rejecting <em>any</em> null hypothesis (Familywise Error Control).</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>How do you decide whether FDR or FEC is a better fit for your analysis?</p>
</div>
</div>
<p>The Bonferroni method provides a bound to familywise error. Here is an exampe of how it works.</p>
<div style="page-break-after: always;"></div>
</section>
</section>
<section id="references" class="level2">
<h2 class="anchored" data-anchor-id="references">References</h2>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>